---
title: "GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles"
author: "Shuye Pu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
   
vignette: >
  %\VignetteIndexEntry{GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

Visualization of next generation sequencing (NGS) data at various genomic features on a genome-wide scale provides a effective way of exploring and communicating experimental results on one hand, yet poses as a tremendous challenge on the other hand, due to the huge amount of data to be processed. Existing software tools like deeptools, ngs.plot and metagene2, while having attractive features and perform reasonably well in relatively simple scenarios, like plotting coverage profiles of fixed genomic loci or regions, have serious limitations in terms of efficiency and flexibility. For instance, deeptools requires 3 steps (3 sub-programs to be run) to generate plots from input files: first, convert .bam files to .bigwig format; second, compute coverage matrix; and last, plot genomic profiles. Huge amount of intermediate data are generated along the way and additional efforts have to be made to integrate these 3 closely related steps. All of them focus on plotting signals within genomic regions or around genomic loci, but not within or around combinations of genomic features. None of them have the capability of performing statistical tests on the data displayed in the profile plots.

To meet the diverse needs of experimental biologists, we have developed GenomicPlot using rich resources available on the R platform (particularly, the Bioconductor). Our GenomicPlot has the following major features:

-   Generates genomic or metagenomic (without introns) plots around gene body and its upstream and downstream regions, the gene body can be further segmented into 5' UTR, CDS and 3' UTR
-   Plots genomic profiles around the start and end of genomic features (like exon or intron), or custom defined genomic regions
-   Plots distance between sample peaks and genomic features, or distance between one set of peaks to another set of peaks
-   All profile line plots have error bands
-   Random features can be generated and plotted to serve as contrast to real features
-   Statistical analysis results on user defined regions are plotted along with the profile plots

## Plot metagene with 5'UTR, CDS and 3'UTR

The lengths of each part of the genes are prorated based on the median length of 5'UTR, CDS and 3'UTR of protein-coding genes in the genome. The total length (including upstream and downstream extensions) are divided into the specified number of bins.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7), fig.ncol=1, fig.sep="\n\n"}
library(GenomicPlot, quietly=TRUE)
txdb <- AnnotationDbi::loadDb(system.file("data", "txdb_chr19.sql", package="GenomicPlotData"))
gf <- prepare_5parts_genomic_features(txdb, meta=TRUE, nbins=100, fiveP=500, threeP=500, 
                                      longest=TRUE)

queryfiles <- system.file("data", "test_clip_chr19.bam", package="GenomicPlotData")
names(queryfiles) <- "query"
inputfiles <- system.file("data", "test_clip_input_chr19.bam", package="GenomicPlotData")
names(inputfiles) <- "input"

handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, 
                          useScore=FALSE, outRle=TRUE, useSizeFactor=TRUE, genome="hg19")

plot_5parts_metagene(queryfiles, gFeatures=gf, inputfiles=inputfiles, scale=FALSE, verbose=FALSE, 
                     transform=FALSE, smooth=TRUE, stranded=TRUE, outPrefix=NULL, 
                     handleInputParams=handleInputParams, heatmap=TRUE, rmOutlier=TRUE, nc=5)

```

## Plot metagene

The lengths of each part of the profile plot are prorated to the lengths of upstream and downstream extensions as well as the median length of protein-coding genes in the genome. The total length (including upstream and downstream extensions) are divided into the specified number of bins.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
gf <- prepare_3parts_genomic_features(txdb, meta=FALSE, nbins=100, fiveP=3000, threeP=2000, 
                                      longest=TRUE)

queryfiles <- system.file("data", "test_chip_chr19.bam", package="GenomicPlotData")
names(queryfiles) <- "query"
inputfiles <- system.file("data", "test_chip_input_chr19.bam", package="GenomicPlotData")
names(inputfiles) <- "input"

handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, 
                          useScore=FALSE, outRle=TRUE, useSizeFactor=TRUE, genome="hg19")

plot_3parts_metagene(queryfiles, gFeatures=gf, inputfiles=inputfiles, scale=FALSE, verbose=FALSE, 
                     transform=FALSE, smooth=TRUE, stranded=TRUE, outPrefix=NULL, 
                     handleInputParams=handleInputParams, heatmap=TRUE, rmOutlier=TRUE, nc=5)

```

## Plot user defined regions

Signal profile in user defined regions provided through a .bed file or narrowPeak file along with heatmaps can be plotted. Multiple samples (.bam files) and multiple set of regions (.bed file) can over-layed on the same figure, or can be output as various combinations. When input files (for ChIPseq and CLIPseq) are available, ratio-over-input are displayed as well.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
centerfiles <- system.file("data", "test_chip_peak_chr19.narrowPeak", package="GenomicPlotData")
 names(centerfiles) <- c("narrowpeak")
 queryfiles <- c(system.file("data", "test_clip_chr19.bam", package="GenomicPlotData"),
                 system.file("data", "test_chip_chr19.bam", package="GenomicPlotData"))
 names(queryfiles) <- c("clip_bam", "chip_bam")
 inputfiles <- c(system.file("data", "test_clip_input_chr19.bam", package="GenomicPlotData"),
                 system.file("data", "test_chip_input_chr19.bam", package="GenomicPlotData"))
 names(inputfiles) <- c("clip_input", "chip_input")

 handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, useScore=FALSE,
                           outRle=TRUE, useSizeFactor=TRUE, genome="hg19")

 plot_reference_region(queryfiles, centerfiles, inputfiles, nbins=100, heatmap=TRUE, scale=FALSE,  
                       regionName="narrowPeak", handleInputParams=handleInputParams, verbose=FALSE, 
                       fiveP=500, threeP=500, smooth=TRUE, transform=FALSE, stranded=TRUE, outPrefix=NULL, 
                       rmOutlier=FALSE, nc=5)
```

## Plot start and end of a genomic feature

When the intron splicing sites are focus of studies, we want to inspect signals at both 5' splicing sites and 3' splicing sites, perhaps want to contrast with the center of introns as well, then displaying all 3 pieces of information in the same panel would be desirable.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
txdb <- AnnotationDbi::loadDb(system.file("data", "txdb_chr19.sql", package="GenomicPlotData"))
queryfiles <- system.file("data", "test_clip_chr19.bam", package="GenomicPlotData")
names(queryfiles) <- "query"
inputfiles <- system.file("data", "test_clip_input_chr19.bam", package="GenomicPlotData")
names(inputfiles) <- "input"
ext <- c(-500, 200, -200, 500)
hl <- c(-50, 50, -50, 50)

handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, useScore=FALSE, 
                          outRle=TRUE, useSizeFactor=TRUE, genome="hg19")
plot_start_end_feature(queryfiles=queryfiles, inputfiles=inputfiles, txdb=txdb, featureName="intron", 
                       binsize=10,handleInputParams=handleInputParams, longest=TRUE, ext=ext, hl=hl, 
                       randomize=TRUE, insert=100, stranded=TRUE, scale=FALSE, smooth=TRUE, 
                       outPrefix=NULL, nc=5)
```

## Plot reference locus 
Difference in signal intensity within specific regions around the reference loci can be tested, and the test statistics can be plotted as boxplot and barplot of mean +/- SE. The test can be done among loci or among samples.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
 centerfiles <- c(system.file("data", "test_clip_peak_chr19.bed", package="GenomicPlotData"),
                  system.file("data", "test_chip_peak_chr19.bed", package="GenomicPlotData"))
 names(centerfiles) <- c("clip_peak", "chip_peak")
 queryfiles <- c(system.file("data", "test_clip_chr19.bam", package="GenomicPlotData"),
                 system.file("data", "test_chip_chr19.bam", package="GenomicPlotData"))
 names(queryfiles) <- c("clip_bam", "chip_bam")
 inputfiles <- c(system.file("data", "test_clip_input_chr19.bam", package="GenomicPlotData"),
                 system.file("data", "test_chip_input_chr19.bam", package="GenomicPlotData"))
 names(inputfiles) <- c("clip_input", "chip_input")

 handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, useScore=FALSE,
                           outRle=TRUE, useSizeFactor=TRUE, genome="hg19")

 plot_reference_locus(queryfiles, centerfiles, ext=c(-500,500), hl=c(-100,100), shade=TRUE, smooth=TRUE,
                     handleInputParams=handleInputParams, binsize=10, refPoint="center", Xlab="Center",
                     inputfiles=inputfiles, stranded=TRUE, scale=FALSE, outPrefix=NULL, verbose=FALSE,
                     transform=FALSE, rmOutlier=FALSE, stats.method="wilcox.test", heatmap=TRUE, nc=5)
```

## Plot reference loci with random sites as contrast

Random sites that deviate from the reference loci within a specified maximum distance can be treated as a control, and the statistical tests will be performed between bona fide and the random sites. Here the reference sites located in different genomic features (5'UTR, CDS, 3'UTR) can tested separately.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
txdb <- AnnotationDbi::loadDb(system.file("data", "txdb_chr19.sql", package="GenomicPlotData"))
 centerfiles <- c(system.file("data", "test_clip_peak_chr19.bed", package="GenomicPlotData"))
 names(centerfiles) <- c("clip_peak")
 queryfiles <- c(system.file("data", "test_clip_chr19.bam", package="GenomicPlotData"))
 names(queryfiles) <- c("clip_bam")
 inputfiles <- c(system.file("data", "test_clip_input_chr19.bam", package="GenomicPlotData"))
 names(inputfiles) <- c("clip_input")

handleInputParams <- list(CLIP_reads=FALSE, fix_width=150, fix_point="start", norm=TRUE, useScore=FALSE,
                          outRle=TRUE, useSizeFactor=TRUE, genome="hg19")

plot_reference_locus_with_random(queryfiles, centerfiles, txdb, ext=c(-500,500), hl=c(-100,100), 
                                 shade=TRUE, handleInputParams=handleInputParams, binsize=10, 
                                 refPoint="center", Xlab="Center", smooth=TRUE,inputfiles=inputfiles, 
                                 stranded=TRUE, scale=FALSE, outPrefix=NULL, verbose=FALSE, 
                                 transform=FALSE, rmOutlier=FALSE, n_random=1, stats.method="wilcox.test", 
                                 nc=5)
```

## Plot peak annotation statistics

Aside from reads coverage profiles, distribution of binding peaks in different gene types and genomic features is also important. Peak annotation statistics are plotted as bar chart for distribution in gene types, and as pie charts for distribution in genomic features. The pie charts are plotted in two different ways: either as percentage of absolute counts or as percentage of feature length-normalized counts in each features. For DNA binding samples, the features, in order of precedence, include "Promoter", "TTS", "5'UTR", "CDS", "3'UTR" and "Intron"; for RNA binding samples, "Promoter" and "TTS" are excluded. "Promoter" is defined as regions 2000 bp upstream of transcription start site (TSS), "TTS" is defined as the region 1000 bp downstream transcription termination sites by default, but these lengths can be adjusted.

```{r, eval=FALSE, fig.show='hold', fig.keep='all', fig.align='center', fig.dim=c(7,7)}
suppressPackageStartupMessages(library(GenomicPlot, quietly=TRUE))
gtffile <- system.file("data", "gencode.v19.annotation_chr19.gtf", package="GenomicPlotData")

centerfile <- system.file("data", "test_chip_peak.narrowPeak", package="GenomicPlotData")
names(centerfile) <- c("narrowPeak")

handleBedparams <- list(fix_width=0, fix_point="center", useScore=FALSE, outRle=FALSE, CLIP_reads=FALSE,
                        norm=FALSE, useSizeFactor=FALSE, genome="hg19")

pa <- plot_peak_annotation(peakfile=centerfile, gtfFile=gtffile, handleInputParams=handleBedparams, 
                           fiveP=2000, threeP=1000, simple=FALSE, RNA=FALSE, verbose=TRUE, 
                           outPrefix=NULL)
```
