<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Shuye Pu" />

<meta name="date" content="2023-02-07" />

<title>GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">GenomicPlot: an R package for efficient and
flexible visualization of NGS coverage profiles</h1>
<h4 class="author">Shuye Pu</h4>
<h4 class="date">2023-02-07</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Visualization of next generation sequencing (NGS) data at various
genomic features on a genome-wide scale provides a effective way of
exploring and communicating experimental results on one hand, yet poses
as a tremendous challenge on the other hand, due to the huge amount of
data to be processed. Existing software tools like deeptools, ngs.plot
and metagene2, while having attractive features and perform reasonably
well in relatively simple scenarios, like plotting coverage profiles of
fixed genomic loci or regions, have serious limitations in terms of
efficiency and flexibility. For instance, deeptools requires 3 steps (3
sub-programs to be run) to generate plots from input files: first,
convert .bam files to .bigwig format; second, compute coverage matrix;
and last, plot genomic profiles. Huge amount of intermediate data are
generated along the way and additional efforts have to be made to
integrate these 3 closely related steps. All of them focus on plotting
signals within genomic regions or around genomic loci, but not within or
around combinations of genomic features. None of them have the
capability of performing statistical tests on the data displayed in the
profile plots.</p>
<p>To meet the diverse needs of experimental biologists, we have
developed GenomicPlot using rich resources available on the R platform
(particularly, the Bioconductor). Our GenomicPlot has the following
major features:</p>
<ul>
<li>Generates genomic or metagenomic (without introns) plots around gene
body and its upstream and downstream regions, the gene body can be
further segmented into 5’ UTR, CDS and 3’ UTR</li>
<li>Plots genomic profiles around the start and end of genomic features
(like exon or intron), or custom defined genomic regions</li>
<li>Plots distance between sample peaks and genomic features, or
distance between one set of peaks to another set of peaks</li>
<li>All profile line plots have error bands</li>
<li>Random features can be generated and plotted to serve as contrast to
real features</li>
<li>Statistical analysis results on user defined regions are plotted
along with the profile plots</li>
</ul>
</div>
<div id="plot-metagene-with-5utr-cds-and-3utr" class="section level2">
<h2>Plot metagene with 5’UTR, CDS and 3’UTR</h2>
<p>The lengths of each part of the genes are prorated based on the
median length of 5’UTR, CDS and 3’UTR of protein-coding genes in the
genome. The total length (including upstream and downstream extensions)
are divided into the specified number of bins.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicPlot, <span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">loadDb</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;txdb_chr19.sql&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">prepare_5parts_genomic_features</span>(txdb, <span class="at">meta=</span><span class="cn">TRUE</span>, <span class="at">nbins=</span><span class="dv">100</span>, <span class="at">fiveP=</span><span class="dv">500</span>, <span class="at">threeP=</span><span class="dv">500</span>, </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">longest=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>queryfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="st">&quot;query&quot;</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>inputfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="st">&quot;input&quot;</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>                          <span class="at">useScore=</span><span class="cn">FALSE</span>, <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_5parts_metagene</span>(queryfiles, <span class="at">gFeatures=</span>gf, <span class="at">inputfiles=</span>inputfiles, <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>, </span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>                     <span class="at">transform=</span><span class="cn">FALSE</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>, <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">outPrefix=</span><span class="cn">NULL</span>, </span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>                     <span class="at">handleInputParams=</span>handleInputParams, <span class="at">heatmap=</span><span class="cn">TRUE</span>, <span class="at">rmOutlier=</span><span class="cn">TRUE</span>, <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-metagene" class="section level2">
<h2>Plot metagene</h2>
<p>The lengths of each part of the profile plot are prorated to the
lengths of upstream and downstream extensions as well as the median
length of protein-coding genes in the genome. The total length
(including upstream and downstream extensions) are divided into the
specified number of bins.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">prepare_3parts_genomic_features</span>(txdb, <span class="at">meta=</span><span class="cn">FALSE</span>, <span class="at">nbins=</span><span class="dv">100</span>, <span class="at">fiveP=</span><span class="dv">3000</span>, <span class="at">threeP=</span><span class="dv">2000</span>, </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">longest=</span><span class="cn">TRUE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>queryfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="st">&quot;query&quot;</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>inputfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="st">&quot;input&quot;</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">useScore=</span><span class="cn">FALSE</span>, <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_3parts_metagene</span>(queryfiles, <span class="at">gFeatures=</span>gf, <span class="at">inputfiles=</span>inputfiles, <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>, </span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>                     <span class="at">transform=</span><span class="cn">FALSE</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>, <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">outPrefix=</span><span class="cn">NULL</span>, </span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>                     <span class="at">handleInputParams=</span>handleInputParams, <span class="at">heatmap=</span><span class="cn">TRUE</span>, <span class="at">rmOutlier=</span><span class="cn">TRUE</span>, <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-user-defined-regions" class="section level2">
<h2>Plot user defined regions</h2>
<p>Signal profile in user defined regions provided through a .bed file
or narrowPeak file along with heatmaps can be plotted. Multiple samples
(.bam files) and multiple set of regions (.bed file) can over-layed on
the same figure, or can be output as various combinations. When input
files (for ChIPseq and CLIPseq) are available, ratio-over-input are
displayed as well.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>centerfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_peak_chr19.narrowPeak&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(centerfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;narrowpeak&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a> queryfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_bam&quot;</span>, <span class="st">&quot;chip_bam&quot;</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a> inputfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>),</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_input&quot;</span>, <span class="st">&quot;chip_input&quot;</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a> handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, <span class="at">useScore=</span><span class="cn">FALSE</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>                           <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_reference_region</span>(queryfiles, centerfiles, inputfiles, <span class="at">nbins=</span><span class="dv">100</span>, <span class="at">heatmap=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">FALSE</span>,  </span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>                       <span class="at">regionName=</span><span class="st">&quot;narrowPeak&quot;</span>, <span class="at">handleInputParams=</span>handleInputParams, <span class="at">verbose=</span><span class="cn">FALSE</span>, </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>                       <span class="at">fiveP=</span><span class="dv">500</span>, <span class="at">threeP=</span><span class="dv">500</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>, <span class="at">transform=</span><span class="cn">FALSE</span>, <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">outPrefix=</span><span class="cn">NULL</span>, </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                       <span class="at">rmOutlier=</span><span class="cn">FALSE</span>, <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-start-and-end-of-a-genomic-feature" class="section level2">
<h2>Plot start and end of a genomic feature</h2>
<p>When the intron splicing sites are focus of studies, we want to
inspect signals at both 5’ splicing sites and 3’ splicing sites, perhaps
want to contrast with the center of introns as well, then displaying all
3 pieces of information in the same panel would be desirable.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">loadDb</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;txdb_chr19.sql&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>queryfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="st">&quot;query&quot;</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>inputfiles <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="st">&quot;input&quot;</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>ext <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">500</span>, <span class="dv">200</span>, <span class="sc">-</span><span class="dv">200</span>, <span class="dv">500</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>hl <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>, <span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, <span class="at">useScore=</span><span class="cn">FALSE</span>, </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_start_end_feature</span>(<span class="at">queryfiles=</span>queryfiles, <span class="at">inputfiles=</span>inputfiles, <span class="at">txdb=</span>txdb, <span class="at">featureName=</span><span class="st">&quot;intron&quot;</span>, </span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                       <span class="at">binsize=</span><span class="dv">10</span>,<span class="at">handleInputParams=</span>handleInputParams, <span class="at">longest=</span><span class="cn">TRUE</span>, <span class="at">ext=</span>ext, <span class="at">hl=</span>hl, </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                       <span class="at">randomize=</span><span class="cn">TRUE</span>, <span class="at">insert=</span><span class="dv">100</span>, <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>, </span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>                       <span class="at">outPrefix=</span><span class="cn">NULL</span>, <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-reference-locus" class="section level2">
<h2>Plot reference locus</h2>
<p>Difference in signal intensity within specific regions around the
reference loci can be tested, and the test statistics can be plotted as
boxplot and barplot of mean +/- SE. The test can be done among loci or
among samples.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a> centerfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_peak_chr19.bed&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>),</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_peak_chr19.bed&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(centerfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_peak&quot;</span>, <span class="st">&quot;chip_peak&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a> queryfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_bam&quot;</span>, <span class="st">&quot;chip_bam&quot;</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a> inputfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>),</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_input&quot;</span>, <span class="st">&quot;chip_input&quot;</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a> handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, <span class="at">useScore=</span><span class="cn">FALSE</span>,</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>                           <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_reference_locus</span>(queryfiles, centerfiles, <span class="at">ext=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">500</span>,<span class="dv">500</span>), <span class="at">hl=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">100</span>,<span class="dv">100</span>), <span class="at">shade=</span><span class="cn">TRUE</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                     <span class="at">handleInputParams=</span>handleInputParams, <span class="at">binsize=</span><span class="dv">10</span>, <span class="at">refPoint=</span><span class="st">&quot;center&quot;</span>, <span class="at">Xlab=</span><span class="st">&quot;Center&quot;</span>,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>                     <span class="at">inputfiles=</span>inputfiles, <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">outPrefix=</span><span class="cn">NULL</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>                     <span class="at">transform=</span><span class="cn">FALSE</span>, <span class="at">rmOutlier=</span><span class="cn">FALSE</span>, <span class="at">stats.method=</span><span class="st">&quot;wilcox.test&quot;</span>, <span class="at">heatmap=</span><span class="cn">TRUE</span>, <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-reference-loci-with-random-sites-as-contrast" class="section level2">
<h2>Plot reference loci with random sites as contrast</h2>
<p>Random sites that deviate from the reference loci within a specified
maximum distance can be treated as a control, and the statistical tests
will be performed between bona fide and the random sites. Here the
reference sites located in different genomic features (5’UTR, CDS,
3’UTR) can tested separately.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">loadDb</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;txdb_chr19.sql&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a> centerfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_peak_chr19.bed&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(centerfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_peak&quot;</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a> queryfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(queryfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_bam&quot;</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a> inputfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_clip_input_chr19.bam&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">names</span>(inputfiles) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clip_input&quot;</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>handleInputParams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CLIP_reads=</span><span class="cn">FALSE</span>, <span class="at">fix_width=</span><span class="dv">150</span>, <span class="at">fix_point=</span><span class="st">&quot;start&quot;</span>, <span class="at">norm=</span><span class="cn">TRUE</span>, <span class="at">useScore=</span><span class="cn">FALSE</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">outRle=</span><span class="cn">TRUE</span>, <span class="at">useSizeFactor=</span><span class="cn">TRUE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_reference_locus_with_random</span>(queryfiles, centerfiles, txdb, <span class="at">ext=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">500</span>,<span class="dv">500</span>), <span class="at">hl=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">100</span>,<span class="dv">100</span>), </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">shade=</span><span class="cn">TRUE</span>, <span class="at">handleInputParams=</span>handleInputParams, <span class="at">binsize=</span><span class="dv">10</span>, </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">refPoint=</span><span class="st">&quot;center&quot;</span>, <span class="at">Xlab=</span><span class="st">&quot;Center&quot;</span>, <span class="at">smooth=</span><span class="cn">TRUE</span>,<span class="at">inputfiles=</span>inputfiles, </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">stranded=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">outPrefix=</span><span class="cn">NULL</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>, </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">transform=</span><span class="cn">FALSE</span>, <span class="at">rmOutlier=</span><span class="cn">FALSE</span>, <span class="at">n_random=</span><span class="dv">1</span>, <span class="at">stats.method=</span><span class="st">&quot;wilcox.test&quot;</span>, </span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">nc=</span><span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="plot-peak-annotation-statistics" class="section level2">
<h2>Plot peak annotation statistics</h2>
<p>Aside from reads coverage profiles, distribution of binding peaks in
different gene types and genomic features is also important. Peak
annotation statistics are plotted as bar chart for distribution in gene
types, and as pie charts for distribution in genomic features. The pie
charts are plotted in two different ways: either as percentage of
absolute counts or as percentage of feature length-normalized counts in
each features. For DNA binding samples, the features, in order of
precedence, include “Promoter”, “TTS”, “5’UTR”, “CDS”, “3’UTR” and
“Intron”; for RNA binding samples, “Promoter” and “TTS” are excluded.
“Promoter” is defined as regions 2000 bp upstream of transcription start
site (TSS), “TTS” is defined as the region 1000 bp downstream
transcription termination sites by default, but these lengths can be
adjusted.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(GenomicPlot, <span class="at">quietly=</span><span class="cn">TRUE</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>gtffile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;gencode.v19.annotation_chr19.gtf&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>centerfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;test_chip_peak.narrowPeak&quot;</span>, <span class="at">package=</span><span class="st">&quot;GenomicPlotData&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(centerfile) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;narrowPeak&quot;</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>handleBedparams <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">fix_width=</span><span class="dv">0</span>, <span class="at">fix_point=</span><span class="st">&quot;center&quot;</span>, <span class="at">useScore=</span><span class="cn">FALSE</span>, <span class="at">outRle=</span><span class="cn">FALSE</span>, <span class="at">CLIP_reads=</span><span class="cn">FALSE</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">norm=</span><span class="cn">FALSE</span>, <span class="at">useSizeFactor=</span><span class="cn">FALSE</span>, <span class="at">genome=</span><span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>pa <span class="ot">&lt;-</span> <span class="fu">plot_peak_annotation</span>(<span class="at">peakfile=</span>centerfile, <span class="at">gtfFile=</span>gtffile, <span class="at">handleInputParams=</span>handleBedparams, </span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>                           <span class="at">fiveP=</span><span class="dv">2000</span>, <span class="at">threeP=</span><span class="dv">1000</span>, <span class="at">simple=</span><span class="cn">FALSE</span>, <span class="at">RNA=</span><span class="cn">FALSE</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>, </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>                           <span class="at">outPrefix=</span><span class="cn">NULL</span>)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
