<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plot profiles of next generation sequencing data in genomic features • GenomicPlot</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Plot profiles of next generation sequencing data in genomic features">
<meta property="og:description" content="Visualization of next generation sequencing (NGS) data is essential for interpreting
   high-throughput genomics experiment results. GenomicPlot facilitates plotting of NGS data in
   various formats (bam, bed, wig and bigwig); both coverage and enrichment over input can be 
   computed and displayed with respect to genomic features (such as UTR, CDS, enhancer), and user
   defined genomic loci or regions. Statistical tests on signal intensity within user defined 
   regions of interest can be performed and represented as boxplots or bar graphs. Parallel 
   processing is used to speed up computation on multicore platforms. In addition to genomic plots
   which is suitable for displaying of coverage of genomic DNA (such as ChIPseq data),
   metagenomic (without introns) plots can also be made for RNAseq or CLIPseq data as well. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/shuye2009/GenomicPlot/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="genomicplot">GenomicPlot<a class="anchor" aria-label="anchor" href="#genomicplot"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The goal of GenomicPlot is to create an efficient visualization tool for next generation sequencing (NGS) data with rich functionality and flexibility. GenomicPlot enables plotting of NGS data in various formats (bam, bed, wig and bigwig); both coverage and enrichment over input can be computed and displayed with respect to genomic features (such as UTR, CDS, enhancer), and user defined genomic loci or regions. Statistical tests on signal intensity within user defined regions of interest can be performed and presented as box plots or pie charts. Parallel processing is enabled to speed up computation on multi-core platforms. In addition to genomic plots which is suitable for displaying of coverage of genomic DNA (such as ChIP-seq data), metagenomic (without introns) plots can also be made for RNA-seq or CLIP-seq data as well. For peak annotation, peaks targeting exonic and intronic regions of genomic features (5’UTR, CDS, 3’UTR) are counted separately.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The following packages are prerequisites:</p>
<p>GenomicRanges (&gt;= 1.46.1), GenomicFeatures, Rsamtools, ggplot2 (&gt;= 3.3.5), tidyr, rtracklayer (&gt;= 1.54.0), plyranges (&gt;= 1.14.0), dplyr (&gt;= 1.0.8), cowplot (&gt;= 1.1.1), VennDiagram, ggplotify, GenomeInfoDb, IRanges, ComplexHeatmap, RCAS (&gt;= 1.20.0), scales (&gt;= 1.2.0), GenomicAlignments (&gt;= 1.30.0), edgeR, forcats, circlize, viridis, ggsignif (&gt;= 0.6.3), ggsci (&gt;= 2.9), genomation (&gt;= 1.26.0), ggpubr</p>
<p>You can install the development version of GenomicPlot from <a href="https://github.com/shuye2009/GenomicPlot" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"shuye2009/GenomicPlot"</span>, build_manual <span class="op">=</span> <span class="cn">TRUE</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Or download the source package from the latest release on <a href="https://github.com/shuye2009/GenomicPlot" class="external-link">GitHub</a> and run in R:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"path-to-source-package/GenomicPlot_x.x.x.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>where “path-to-source-package” is the absolute path to the file “GenomicPlot_x.x.x.tar.gz”, substitute ‘x’ with the version number of your downloaded package ## Examples</p>
<p>The following is a basic example which shows you how to visualize ChIP-seq peaks and iCLIP-seq peaks in different parts of genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/shuye2009/GenomicPlot" class="external-link">GenomicPlot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/prepare_5parts_genomic_features.html">prepare_5parts_genomic_features</a></span><span class="op">(</span><span class="va">txdb</span>, meta <span class="op">=</span> <span class="cn">TRUE</span>, nbins <span class="op">=</span> <span class="fl">100</span>, fiveP <span class="op">=</span> <span class="op">-</span><span class="fl">2000</span>, threeP <span class="op">=</span> <span class="fl">1000</span>, longest <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.narrowPeak"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.bed"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_clip_peak_chr19.bed"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NarrowPeak"</span>, <span class="st">"SummitPeak"</span>, <span class="st">"iCLIPPeak"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">op</span> <span class="op">&lt;-</span> <span class="st">"test_plot_5parts_metagene"</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0</span>, fix_width <span class="op">=</span> <span class="fl">100</span>, fix_point <span class="op">=</span> <span class="st">"center"</span>, norm <span class="op">=</span> <span class="cn">FALSE</span>, useScore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          outRle <span class="op">=</span> <span class="cn">TRUE</span>, useSizeFactor <span class="op">=</span> <span class="cn">FALSE</span>, genome <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_5parts_metagene.html">plot_5parts_metagene</a></span><span class="op">(</span>queryFiles <span class="op">=</span> <span class="va">queryfiles</span>, </span>
<span>                     gFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>metaF <span class="op">=</span> <span class="va">gf</span><span class="op">)</span>, </span>
<span>                     inputFiles <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                     importParams <span class="op">=</span> <span class="va">importParams</span>,</span>
<span>                     verbose <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     smooth <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     scale <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     stranded <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     outPrefix <span class="op">=</span> <span class="va">op</span>, </span>
<span>                     transform <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>                     heatmap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     rmOutlier <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                     heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     nc <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span>                     </span></code></pre></div>
<p><img src="./inst/tests/test_output/test_plot_5parts_metagene1_1.png" alt="image1"><img src="./inst/tests/test_output/test_plot_5parts_metagene1_2.png" alt="image1"><img src="./inst/tests/test_output/test_plot_5parts_metagene1_3.png" alt="image1"></p>
<p>The following example shows you how to visualize distance between iCLIP_peaks/ChIP-seq narrowPeaks and ChIP-seq summit Peaks.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">op</span> <span class="op">&lt;-</span> <span class="st">"test_plot_locus"</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0</span>, fix_width <span class="op">=</span> <span class="fl">0</span>, fix_point <span class="op">=</span> <span class="st">"center"</span>, norm <span class="op">=</span> <span class="cn">FALSE</span>, useScore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          outRle <span class="op">=</span> <span class="cn">TRUE</span>, useSizeFactor <span class="op">=</span> <span class="cn">FALSE</span>, genome <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_locus.html">plot_locus</a></span><span class="op">(</span>queryFiles <span class="op">=</span> <span class="va">queryfiles</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                     centerFiles <span class="op">=</span> <span class="va">queryfiles</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>                     ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1000</span>, <span class="fl">1000</span><span class="op">)</span>, </span>
<span>                     hl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span>, </span>
<span>                     inputFiles <span class="op">=</span> <span class="cn">NULL</span>,                              </span>
<span>                     importParams <span class="op">=</span> <span class="va">importParams</span>, </span>
<span>                     shade <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     binSize <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                     refPoint <span class="op">=</span> <span class="st">"center"</span>, </span>
<span>                     Xlab <span class="op">=</span> <span class="st">"Summit"</span>,</span>
<span>                     verbose <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     smooth <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     scale <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     stranded <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     outPrefix <span class="op">=</span> <span class="va">op</span>, </span>
<span>                     transform <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>                     heatmap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     rmOutlier <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                     nc <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span>                     </span></code></pre></div>
<p><img src="./inst/tests/test_output/test_plot_locus1_1.png" alt="image1"><img src="./inst/tests/test_output/test_plot_locus1_2.png" alt="image1"><img src="./inst/tests/test_output/test_plot_locus1_3.png" alt="image1"><img src="./inst/tests/test_output/test_plot_locus1_4.png" alt="image1"></p>
<p>The following example shows you how to annotate ChIP-seq peaks. The annotation statistics is shown in “test_plot_peak_annotation.png”. The detailed annotation is in the table “summitPeak_targeted_annotated_gene.tab”.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">op</span> <span class="op">&lt;-</span> <span class="st">"test_plot_peak_annotation"</span></span>
<span></span>
<span><span class="va">gtffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gencode.v19.annotation_chr19.gtf"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0</span>, fix_width <span class="op">=</span> <span class="fl">21</span>, fix_point <span class="op">=</span> <span class="st">"center"</span>, norm <span class="op">=</span> <span class="cn">FALSE</span>, useScore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          outRle <span class="op">=</span> <span class="cn">FALSE</span>, useSizeFactor <span class="op">=</span> <span class="cn">FALSE</span>, genome <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span></span>
<span>                          </span>
<span><span class="fu"><a href="reference/plot_peak_annotation.html">plot_peak_annotation</a></span><span class="op">(</span>peakFile <span class="op">=</span> <span class="va">queryfiles</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>                     gtfFile <span class="op">=</span> <span class="va">gtffile</span>, </span>
<span>                     importParams <span class="op">=</span> <span class="va">importParams</span>, </span>
<span>                     fiveP <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span>, </span>
<span>                     threeP <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                     outPrefix <span class="op">=</span> <span class="va">op</span>, </span>
<span>                     verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="./inst/tests/test_output/test_plot_peak_annotation1_1.png" alt="image1"><img src="./inst/tests/test_output/test_plot_peak_annotation1_2.png" alt="image1"><img src="./inst/tests/test_output/test_plot_peak_annotation1_3.png" alt="image1"><img src="./inst/tests/test_output/test_plot_peak_annotation1_4.png" alt="image1"></p>
<p>For more examples, please check out the package vignette using browseVignettes(“GenomicPlot”).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/shuye2009/GenomicPlot/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/shuye2009/GenomicPlot/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GenomicPlot</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Shuye Pu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6664-8438" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
