<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles • GenomicPlot</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles">
<meta property="og:description" content="GenomicPlot">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GenomicPlot: an R package for efficient and
flexible visualization of NGS coverage profiles</h1>
                        <h4 data-toc-skip class="author">Shuye Pu</h4>
            
            <h4 data-toc-skip class="date">2023-05-08</h4>
      
      
      <div class="hidden name"><code>GenomicPlot_vignettes.rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Visualization of next generation sequencing (NGS) data at various
genomic features on a genome-wide scale provides an effective way of
exploring and communicating experimental results on one hand, yet poses
as a tremendous challenge on the other hand, due to the huge amount of
data to be processed. Existing software tools like deeptools, ngs.plot,
coverageView and metagene2, while having attractive features and perform
reasonably well in relatively simple scenarios, like plotting coverage
profiles of fixed genomic loci or regions, have serious limitations in
terms of efficiency and flexibility. For instance, deeptools requires 3
steps (3 sub-programs to be run) to generate plots from input files:
first, convert .bam files to .bigwig format; second, compute coverage
matrix; and last, plot genomic profiles. Huge amount of intermediate
data are generated along the way and additional efforts have to be made
to integrate these 3 closely related steps. All of them focus on
plotting signals within genomic regions or around genomic loci, but not
within or around combinations of genomic features. None of them have the
capability of performing statistical tests on the data displayed in the
profile plots.</p>
<p>To meet the diverse needs of experimental biologists, we have
developed GenomicPlot using rich resources available on the R platform
(particularly, the Bioconductor). Our GenomicPlot has the following
major features:</p>
<ul>
<li>Generating genomic (with introns) or metagenomic (without introns)
plots around gene body and its upstream and downstream regions, the gene
body can be further segmented into 5’ UTR, CDS and 3’ UTR</li>
<li>Plotting genomic profiles around the start and end of genomic
features (like exons or introns), or custom defined genomic regions</li>
<li>Plotting distance between sample peaks and genomic features, or
distance from one set of peaks to another set of peaks</li>
<li>Plotting peak annotation statistics (distribution in different type
of genes, and in different parts of genes)</li>
<li>Plotting peak overlaps as Venn diagrams</li>
<li>All profile line plots have error bands</li>
<li>Random features can be generated and plotted to serve as contrast to
real features</li>
<li>Statistical analysis results on user defined regions are plotted
along with the profile plots</li>
</ul>
</div>
<div class="section level2">
<h2 id="plot-genemetagene-with-5utr-cds-and-3utr">Plot gene/metagene with 5’UTR, CDS and 3’UTR<a class="anchor" aria-label="anchor" href="#plot-genemetagene-with-5utr-cds-and-3utr"></a>
</h2>
<p>The lengths of each part of the genes are prorated based on the
median length of 5’UTR, CDS and 3’UTR of protein-coding genes in the
genome. The total length (including upstream and downstream extensions)
are divided into the specified number of bins. Subsets of genes can be
plotted as overlays for comparison.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GenomicPlot</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_5parts_genomic_features.html">prepare_5parts_genomic_features</a></span><span class="op">(</span><span class="va">txdb</span>, meta<span class="op">=</span><span class="cn">TRUE</span>, nbins<span class="op">=</span><span class="fl">100</span>, fiveP<span class="op">=</span><span class="op">-</span><span class="fl">2000</span>, threeP<span class="op">=</span><span class="fl">1000</span>, </span>
<span>                                      longest<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clip_bam"</span></span>
<span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clip_input"</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, fix_width<span class="op">=</span><span class="fl">0</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                          useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_5parts_metagene.html">plot_5parts_metagene</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>                     gFeatures_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"metagene"</span><span class="op">=</span><span class="va">gf</span><span class="op">)</span>, </span>
<span>                     inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>                     scale<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     transform<span class="op">=</span><span class="cn">NA</span>, </span>
<span>                     smooth<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>                     importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>                     heatmap<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     rmOutlier<span class="op">=</span><span class="fl">0</span>, </span>
<span>                     nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_5parts_metagene2_1.png" alt="Individual sample profile and heatmap" width="75%"><p class="caption">
Individual sample profile and heatmap
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_5parts_metagene2_2.png" alt="Profile overlay" width="75%"><p class="caption">
Profile overlay
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_5parts_metagene2_3.png" alt="Ratio-over-input profile and heatmap" width="75%"><p class="caption">
Ratio-over-input profile and heatmap
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-genemetagene-body">Plot gene/metagene body<a class="anchor" aria-label="anchor" href="#plot-genemetagene-body"></a>
</h2>
<p>The lengths of each part of the profile plot are prorated to the
lengths of upstream and downstream extensions as well as the median
length of protein-coding genes in the genome. The total length
(including upstream and downstream extensions) are divided into the
specified number of bins.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_3parts_genomic_features.html">prepare_3parts_genomic_features</a></span><span class="op">(</span><span class="va">txdb</span>, meta<span class="op">=</span><span class="cn">FALSE</span>, nbins<span class="op">=</span><span class="fl">100</span>, fiveP<span class="op">=</span><span class="op">-</span><span class="fl">3000</span>, threeP<span class="op">=</span><span class="fl">2000</span>, </span>
<span>                                      longest<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chip_bam"</span></span>
<span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chip_input"</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                          useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">TRUE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_3parts_metagene.html">plot_3parts_metagene</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>                     gFeatures<span class="op">=</span><span class="va">gf</span>, </span>
<span>                     inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>                     scale<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     transform<span class="op">=</span><span class="cn">NA</span>, </span>
<span>                     smooth<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>                     importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>                     heatmap<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     rmOutlier<span class="op">=</span><span class="fl">0</span>, </span>
<span>                     nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_3parts_metagene_1.png" alt="Individual sample profile and heatmap" width="75%"><p class="caption">
Individual sample profile and heatmap
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_3parts_metagene_2.png" alt="Profile overlay" width="75%"><p class="caption">
Profile overlay
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_3parts_metagene_3.png" alt="Ratio-over-input profile and heatmap" width="75%"><p class="caption">
Ratio-over-input profile and heatmap
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-along-the-ranges-of-genomic-features">Plot along the ranges of genomic features<a class="anchor" aria-label="anchor" href="#plot-along-the-ranges-of-genomic-features"></a>
</h2>
<p>Signal profile along with heatmaps in genomic features or user
defined genomic regions provided through a .bed file or narrowPeak file
can be plotted. Multiple samples (.bam files) and multiple set of
regions (.bed file) can be overlayed on the same figure, or can be
output as various combinations. When input files (for input samples) are
available, ratio-over-input are displayed as well. Statistical
comparisons between samples or between features can be plotted as
boxplots or barplots of means 0b1 S.E.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.narrowPeak"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NarrowPeak"</span><span class="op">)</span></span>
<span> <span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_bam"</span>, <span class="st">"chip_bam"</span><span class="op">)</span></span>
<span> <span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_input"</span>, <span class="st">"chip_input"</span><span class="op">)</span></span>
<span></span>
<span> <span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                           useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span> <span class="fu"><a href="../reference/plot_region.html">plot_region</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>             centerFiles<span class="op">=</span><span class="va">centerfiles</span>, </span>
<span>             inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>             nbins<span class="op">=</span><span class="fl">100</span>, </span>
<span>             heatmap<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>             scale<span class="op">=</span><span class="cn">FALSE</span>,  </span>
<span>             regionName<span class="op">=</span><span class="st">"narrowPeak"</span>, </span>
<span>             importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>             verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>             fiveP<span class="op">=</span><span class="op">-</span><span class="fl">500</span>, </span>
<span>             threeP<span class="op">=</span><span class="fl">500</span>, </span>
<span>             smooth<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>             transform<span class="op">=</span><span class="cn">NA</span>, </span>
<span>             stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>             outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>             rmOutlier<span class="op">=</span><span class="fl">0</span>, </span>
<span>             nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_21.png" alt="Individual sample profile" width="75%"><p class="caption">
Individual sample profile
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_22.png" alt="Individual sample coverage stats in narrowPeak" width="100%"><p class="caption">
Individual sample coverage stats in narrowPeak
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_23.png" alt="Individual sample coverage profiles and heatmaps in narrowPeak" width="100%"><p class="caption">
Individual sample coverage profiles and heatmaps in narrowPeak
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_24.png" alt="Ratio-over-input profile" width="75%"><p class="caption">
Ratio-over-input profile
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_25.png" alt="Ratio-over-input stats" width="75%"><p class="caption">
Ratio-over-input stats
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_26.png" alt="Ratio-over-input profiles and heatmaps" width="75%"><p class="caption">
Ratio-over-input profiles and heatmaps
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_region_27.png" alt="Program names and arguments" width="75%"><p class="caption">
Program names and arguments
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-start-end-and-center-of-a-genomic-feature">Plot start, end and center of a genomic feature<a class="anchor" aria-label="anchor" href="#plot-start-end-and-center-of-a-genomic-feature"></a>
</h2>
<p>When the intron splicing sites are focus of studies, we want to
inspect signals at both 5’ splicing sites and 3’ splicing sites, perhaps
want to contrast with the center of introns as well, then displaying all
3 pieces of information in the same panel would be desirable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clip_bam"</span></span>
<span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clip_input"</span></span>
<span><span class="va">ext</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">500</span>, <span class="fl">200</span>, <span class="op">-</span><span class="fl">200</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">hl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span>, <span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, fix_width<span class="op">=</span><span class="fl">0</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                          useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_start_end.html">plot_start_end</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>               inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>               txdb<span class="op">=</span><span class="va">txdb</span>, </span>
<span>               centerFiles<span class="op">=</span><span class="st">"intron"</span>, </span>
<span>               binSize<span class="op">=</span><span class="fl">10</span>,</span>
<span>               importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>               ext<span class="op">=</span><span class="va">ext</span>, </span>
<span>               hl<span class="op">=</span><span class="va">hl</span>, </span>
<span>               insert<span class="op">=</span><span class="fl">100</span>, </span>
<span>               stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>               scale<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>               smooth<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>               outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>               nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_stat_end_3.png" alt="Query and input sample profiles in start, center and end of intron" width="75%"><p class="caption">
Query and input sample profiles in start, center and end of intron
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_stat_end_4.png" alt="Ratio-over profile in start, center and end of intron" width="75%"><p class="caption">
Ratio-over profile in start, center and end of intron
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-genomic-loci-start-end-or-center-of-a-feature">Plot genomic loci (start, end or center of a feature)<a class="anchor" aria-label="anchor" href="#plot-genomic-loci-start-end-or-center-of-a-feature"></a>
</h2>
<p>Difference in signal intensity within specific regions around the
reference loci can be tested, and the test statistics can be plotted as
boxplot and barplot of mean +/- SE. The test can be done among loci or
among samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_clip_peak_chr19.bed"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.bed"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iCLIPPeak"</span>, <span class="st">"SummitPeak"</span><span class="op">)</span></span>
<span> <span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_bam"</span>, <span class="st">"chip_bam"</span><span class="op">)</span></span>
<span> <span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_input"</span>, <span class="st">"chip_input"</span><span class="op">)</span></span>
<span></span>
<span> <span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                           useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">TRUE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span> <span class="fu"><a href="../reference/plot_locus.html">plot_locus</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>            centerFiles<span class="op">=</span><span class="va">centerfiles</span>, </span>
<span>            ext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">500</span>,<span class="fl">500</span><span class="op">)</span>, </span>
<span>            hl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span>, </span>
<span>            shade<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>            smooth<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>            importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>            binSize<span class="op">=</span><span class="fl">10</span>, </span>
<span>            refPoint<span class="op">=</span><span class="st">"center"</span>, </span>
<span>            Xlab<span class="op">=</span><span class="st">"Center"</span>, </span>
<span>            inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>            stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>            scale<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>            verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            transform<span class="op">=</span><span class="cn">NA</span>, </span>
<span>            rmOutlier<span class="op">=</span><span class="fl">0</span>, </span>
<span>            statsMethod<span class="op">=</span><span class="st">"wilcox.test"</span>, </span>
<span>            heatmap<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>            nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_21.png" alt="Ratio-over-input for clip signal around clip and chip peaks" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip and chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_22.png" alt="Ratio-over-input for clip signal around clip and chip peaks" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip and chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_23.png" alt="Ratio-over-input for chip signal around clip and chip peaks" width="75%"><p class="caption">
Ratio-over-input for chip signal around clip and chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_24.png" alt="Ratio-over-input for chip signal around clip and chip peaks" width="75%"><p class="caption">
Ratio-over-input for chip signal around clip and chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_25.png" alt="Ratio-over-input for chip and clip signal around clip peaks" width="75%"><p class="caption">
Ratio-over-input for chip and clip signal around clip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_26.png" alt="Ratio-over-input for chip and clip signal around clip peaks" width="75%"><p class="caption">
Ratio-over-input for chip and clip signal around clip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_27.png" alt="Ratio-over-input for chip and clip signal around chip peaks" width="75%"><p class="caption">
Ratio-over-input for chip and clip signal around chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_28.png" alt="Ratio-over-input for chip and clip signal around chip peaks" width="75%"><p class="caption">
Ratio-over-input for chip and clip signal around chip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_29.png" alt="Ratio-over-input profiles for chip and clip signal around clip and chip peaks" width="75%"><p class="caption">
Ratio-over-input profiles for chip and clip signal around clip and chip
peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus2_30.png" alt="Ratio-over-input profiles and heatmap for chip and clip signal" width="100%"><p class="caption">
Ratio-over-input profiles and heatmap for chip and clip signal
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-genomic-loci-with-random-sites-as-contrast">Plot genomic loci with random sites as contrast<a class="anchor" aria-label="anchor" href="#plot-genomic-loci-with-random-sites-as-contrast"></a>
</h2>
<p>Random sites that deviate from the reference loci within a specified
maximum distance can be treated as a control, and the statistical tests
will be performed between bona fide and the random sites. Here the
reference sites located in different genomic features (5’UTR, CDS,
3’UTR) are tested separately. To avoid sampling bias, increase
“n_random”.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_clip_peak_chr19.bed"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iCLIPPeak"</span><span class="op">)</span></span>
<span> <span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_bam"</span><span class="op">)</span></span>
<span> <span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip_input"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                          useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">TRUE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_locus_with_random.html">plot_locus_with_random</a></span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, </span>
<span>                       centerFiles<span class="op">=</span><span class="va">centerfiles</span>, </span>
<span>                       <span class="va">txdb</span>, </span>
<span>                       ext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">500</span>,<span class="fl">500</span><span class="op">)</span>, </span>
<span>                       hl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span>, </span>
<span>                       shade<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                       importParams<span class="op">=</span><span class="va">importParams</span>, </span>
<span>                       binSize<span class="op">=</span><span class="fl">10</span>, </span>
<span>                       refPoint<span class="op">=</span><span class="st">"center"</span>, </span>
<span>                       Xlab<span class="op">=</span><span class="st">"Center"</span>, </span>
<span>                       smooth<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span>
<span>                       stranded<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                       scale<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                       outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span>
<span>                       verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                       transform<span class="op">=</span><span class="cn">NA</span>, </span>
<span>                       rmOutlier<span class="op">=</span><span class="fl">0</span>, </span>
<span>                       n_random<span class="op">=</span><span class="fl">1</span>, </span>
<span>                       statsMethod<span class="op">=</span><span class="st">"wilcox.test"</span>, </span>
<span>                       nc<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus_with_random_27.png" alt="Ratio-over-input for clip signal around clip peaks in 5'UTR" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip peaks in 5’UTR
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus_with_random_28.png" alt="Ratio-over-input for clip signal around clip peaks in 5'UTR" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip peaks in 5’UTR
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus_with_random_35.png" alt="Ratio-over-input for clip signal around clip peaks" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip peaks
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_locus_with_random_36.png" alt="Ratio-over-input for clip signal around clip peaks" width="75%"><p class="caption">
Ratio-over-input for clip signal around clip peaks
</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-peak-annotation-statistics">Plot peak annotation statistics<a class="anchor" aria-label="anchor" href="#plot-peak-annotation-statistics"></a>
</h2>
<p>Aside from reads coverage profiles, distribution of binding peaks in
different gene types and genomic features is also important. Peak
annotation statistics are plotted as bar chart for distribution in gene
types, and as pie charts for distribution in genomic features. The pie
charts are plotted in two different ways: either as percentage of
absolute counts or as percentage of feature length-normalized counts in
each features. For DNA binding samples, the features, in order of
precedence, include “Promoter”, “TTS”, “5’UTR”, “CDS”, “3’UTR” and
“Intron”; for RNA binding samples, “Promoter” and “TTS” are excluded.
“Promoter” is defined as regions 2000 bp upstream of transcription start
site (TSS) and 300 bp downstream TSS, “TTS” is defined as the region
1000 bp downstream cleavage site or the length between cleavage site and
the start of the next gene, whichever is shorter, but these lengths can
be adjusted. To save annotation results (both peak-oriented and
gene-oriented), set “verbose”=TRUE.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gtffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gencode.v19.annotation_chr19.gtf"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">centerfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.bed"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfile</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SummitPeak"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">handleBedparams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fix_width<span class="op">=</span><span class="fl">0</span>, fix_point<span class="op">=</span><span class="st">"center"</span>, useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                        offset<span class="op">=</span><span class="fl">0</span>, norm<span class="op">=</span><span class="cn">FALSE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_peak_annotation.html">plot_peak_annotation</a></span><span class="op">(</span>peakFile<span class="op">=</span><span class="va">centerfile</span>, </span>
<span>                           gtfFile<span class="op">=</span><span class="va">gtffile</span>, </span>
<span>                           importParams<span class="op">=</span><span class="va">handleBedparams</span>, </span>
<span>                           fiveP<span class="op">=</span><span class="op">-</span><span class="fl">2000</span>, </span>
<span>                           dsTSS<span class="op">=</span><span class="fl">300</span>, </span>
<span>                           threeP<span class="op">=</span><span class="fl">1000</span>, </span>
<span>                           simple<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                           verbose<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                           outPrefix<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../tests/test_output/test_plot_peak_annotation1_1.png" alt="Distribution of chip peak in various types of gene" width="75%"><p class="caption">
Distribution of chip peak in various types of gene
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_peak_annotation1_2.png" alt="Distribution of chip peak in various parts of gene" width="75%"><p class="caption">
Distribution of chip peak in various parts of gene
</p>
</div>
<div class="figure">
<img src="../tests/test_output/test_plot_peak_annotation1_3.png" alt="Density of chip peak in various parts of gene" width="75%"><p class="caption">
Density of chip peak in various parts of gene
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
