<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/shuye2009/GenomicPlot/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="aov_TukeyHSD.html">aov_TukeyHSD()</a></code> </p>
        </td>
        <td><p>Perform one-way ANOVA and post hoc TukeyHSD tests</p></td>
      </tr><tr><td>
          <p><code><a href="check_constraints.html">check_constraints()</a></code> </p>
        </td>
        <td><p>Check constraints of genomic ranges</p></td>
      </tr><tr><td>
          <p><code><a href="draw_boxplot_by_factor.html">draw_boxplot_by_factor()</a></code> </p>
        </td>
        <td><p>Plot boxplot with two factors</p></td>
      </tr><tr><td>
          <p><code><a href="draw_boxplot_wo_outlier.html">draw_boxplot_wo_outlier()</a></code> </p>
        </td>
        <td><p>Plot boxplot without outliers</p></td>
      </tr><tr><td>
          <p><code><a href="draw_combo_plot.html">draw_combo_plot()</a></code> </p>
        </td>
        <td><p>Make combo plot for statistics plots</p></td>
      </tr><tr><td>
          <p><code><a href="draw_locus_profile.html">draw_locus_profile()</a></code> </p>
        </td>
        <td><p>Plot signal profile around genomic loci</p></td>
      </tr><tr><td>
          <p><code><a href="draw_matrix_heatmap.html">draw_matrix_heatmap()</a></code> </p>
        </td>
        <td><p>Display matrix as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="draw_mean_se_barplot.html">draw_mean_se_barplot()</a></code> </p>
        </td>
        <td><p>Plot barplot for mean with standard error bars</p></td>
      </tr><tr><td>
          <p><code><a href="draw_quantile_plot.html">draw_quantile_plot()</a></code> </p>
        </td>
        <td><p>Plot quantile over value</p></td>
      </tr><tr><td>
          <p><code><a href="draw_rank_plot.html">draw_rank_plot()</a></code> </p>
        </td>
        <td><p>Plot fraction of cumulative sum over rank</p></td>
      </tr><tr><td>
          <p><code><a href="draw_region_landmark.html">draw_region_landmark()</a></code> </p>
        </td>
        <td><p>Plot genomic region landmark indicator</p></td>
      </tr><tr><td>
          <p><code><a href="draw_region_name.html">draw_region_name()</a></code> </p>
        </td>
        <td><p>Plot genomic region names</p></td>
      </tr><tr><td>
          <p><code><a href="draw_region_profile.html">draw_region_profile()</a></code> </p>
        </td>
        <td><p>Plot signal profile in genomic regions</p></td>
      </tr><tr><td>
          <p><code><a href="draw_stacked_profile.html">draw_stacked_profile()</a></code> </p>
        </td>
        <td><p>Plot signal profile around start, center, and end of genomic regions</p></td>
      </tr><tr><td>
          <p><code><a href="effective_size.html">effective_size()</a></code> </p>
        </td>
        <td><p>Normalize sample library size to effective size</p></td>
      </tr><tr><td>
          <p><code><a href="extract_longest_tx.html">extract_longest_tx()</a></code> </p>
        </td>
        <td><p>Extract the longest transcript for each protein-coding genes</p></td>
      </tr><tr><td>
          <p><code><a href="filter_by_nonoverlaps_stranded.html">filter_by_nonoverlaps_stranded()</a></code> </p>
        </td>
        <td><p>Filter GRanges by nonoverlaps in stranded way</p></td>
      </tr><tr><td>
          <p><code><a href="filter_by_overlaps_stranded.html">filter_by_overlaps_stranded()</a></code> </p>
        </td>
        <td><p>Filter GRanges by overlaps in a stranded way</p></td>
      </tr><tr><td>
          <p><code><a href="gene2tx.html">gene2tx()</a></code> </p>
        </td>
        <td><p>Translate gene names to transcript ids using a GTF file for a subset of genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_genomic_feature_coordinates.html">get_genomic_feature_coordinates()</a></code> </p>
        </td>
        <td><p>Extract genomic features from TxDb object</p></td>
      </tr><tr><td>
          <p><code><a href="get_targeted_genes.html">get_targeted_genes()</a></code> </p>
        </td>
        <td><p>Get the number of peaks overlapping each feature of all protein-coding genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_txdb_features.html">get_txdb_features()</a></code> </p>
        </td>
        <td><p>Get genomic coordinates of features of protein-coding genes</p></td>
      </tr><tr><td>
          <p><code><a href="gr2df.html">gr2df()</a></code> </p>
        </td>
        <td><p>Convert GRanges to dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="handle_bam.html">handle_bam()</a></code> </p>
        </td>
        <td><p>Handle files in bam format</p></td>
      </tr><tr><td>
          <p><code><a href="handle_bed.html">handle_bed()</a></code> </p>
        </td>
        <td><p>Handle files in bed|narrowPeak|broadPeak format</p></td>
      </tr><tr><td>
          <p><code><a href="handle_bw.html">handle_bw()</a></code> </p>
        </td>
        <td><p>Handle files in bw|bigwig|bigWig|BigWig|BW|BIGWIG format</p></td>
      </tr><tr><td>
          <p><code><a href="handle_input.html">handle_input()</a></code> </p>
        </td>
        <td><p>Hand input of NGS data with various formats</p></td>
      </tr><tr><td>
          <p><code><a href="handle_wig.html">handle_wig()</a></code> </p>
        </td>
        <td><p>Handle files in wig format</p></td>
      </tr><tr><td>
          <p><code><a href="make_subTxDb_from_GTF.html">make_subTxDb_from_GTF()</a></code> </p>
        </td>
        <td><p>Make TxDb object from a GTF file for a subset of genes</p></td>
      </tr><tr><td>
          <p><code><a href="overlap_pair.html">overlap_pair()</a></code> </p>
        </td>
        <td><p>Plot two-sets Venn diagram</p></td>
      </tr><tr><td>
          <p><code><a href="overlap_quad.html">overlap_quad()</a></code> </p>
        </td>
        <td><p>Plot four-sets Venn diagram</p></td>
      </tr><tr><td>
          <p><code><a href="overlap_triple.html">overlap_triple()</a></code> </p>
        </td>
        <td><p>Plot three-sets Venn diagram</p></td>
      </tr><tr><td>
          <p><code><a href="parallel_countOverlaps.html">parallel_countOverlaps()</a></code> </p>
        </td>
        <td><p>Parallel execution of countOverlaps</p></td>
      </tr><tr><td>
          <p><code><a href="parallel_scoreMatrixBin.html">parallel_scoreMatrixBin()</a></code> </p>
        </td>
        <td><p>Parallel execution of scoreMatrixBin on a huge target windows object split into chunks</p></td>
      </tr><tr><td>
          <p><code><a href="plot_3parts_metagene.html">plot_3parts_metagene()</a></code> </p>
        </td>
        <td><p>Plot promoter, gene body and TTS</p></td>
      </tr><tr><td>
          <p><code><a href="plot_5parts_metagene.html">plot_5parts_metagene()</a></code> </p>
        </td>
        <td><p>Plot promoter, 5'UTR, CDS, 3'UTR and TTS</p></td>
      </tr><tr><td>
          <p><code><a href="plot_bam_correlation.html">plot_bam_correlation()</a></code> </p>
        </td>
        <td><p>Plot correlation of bam files</p></td>
      </tr><tr><td>
          <p><code><a href="plot_locus.html">plot_locus()</a></code> </p>
        </td>
        <td><p>Plot signal around custom genomic loci</p></td>
      </tr><tr><td>
          <p><code><a href="plot_locus_with_random.html">plot_locus_with_random()</a></code> </p>
        </td>
        <td><p>Plot signal around custom genomic loci and random loci for comparison</p></td>
      </tr><tr><td>
          <p><code><a href="plot_overlap_bed.html">plot_overlap_bed()</a></code> </p>
        </td>
        <td><p>Plot Venn diagrams depicting overlap of genomic regions</p></td>
      </tr><tr><td>
          <p><code><a href="plot_overlap_genes.html">plot_overlap_genes()</a></code> </p>
        </td>
        <td><p>Plot Venn diagrams depicting overlap of gene lists</p></td>
      </tr><tr><td>
          <p><code><a href="plot_peak_annotation.html">plot_peak_annotation()</a></code> </p>
        </td>
        <td><p>Annotate peaks with genomic features and genes</p></td>
      </tr><tr><td>
          <p><code><a href="plot_region.html">plot_region()</a></code> </p>
        </td>
        <td><p>Plot signal inside as well as around custom genomic regions</p></td>
      </tr><tr><td>
          <p><code><a href="plot_start_end.html">plot_start_end()</a></code> </p>
        </td>
        <td><p>Plot signals around the start and the end of genomic features</p></td>
      </tr><tr><td>
          <p><code><a href="plot_start_end_with_random.html">plot_start_end_with_random()</a></code> </p>
        </td>
        <td><p>Plot signals around the start and the end of genomic features and random regions</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_3parts_genomic_features.html">prepare_3parts_genomic_features()</a></code> </p>
        </td>
        <td><p>Demarcate genes into promoter, gene body  and TTS features</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_5parts_genomic_features.html">prepare_5parts_genomic_features()</a></code> </p>
        </td>
        <td><p>Demarcate genes into promoter, 5'UTR, CDS, 3'UTR and TTS features</p></td>
      </tr><tr><td>
          <p><code><a href="process_scoreMatrix.html">process_scoreMatrix()</a></code> </p>
        </td>
        <td><p>Preprocess scoreMatrix before plotting</p></td>
      </tr><tr><td>
          <p><code><a href="rank_rows.html">rank_rows()</a></code> </p>
        </td>
        <td><p>Rank rows of a matrix based on user input</p></td>
      </tr><tr><td>
          <p><code><a href="rm_outlier.html">rm_outlier()</a></code> </p>
        </td>
        <td><p>Remove outliers from scoreMatrix</p></td>
      </tr><tr><td>
          <p><code><a href="start_parallel.html">start_parallel()</a></code> </p>
        </td>
        <td><p>Prepare for parallel processing</p></td>
      </tr><tr><td>
          <p><code><a href="stop_parallel.html">stop_parallel()</a></code> </p>
        </td>
        <td><p>Stop parallel processing</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

