<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot signals around the start and the end of genomic features — plot_start_end • GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot signals around the start and the end of genomic features — plot_start_end"><meta property="og:description" content="Plot reads or peak Coverage/base/gene of samples given in the query files around start and end of custom features. The upstream and downstream windows
can be given separately, within the window, a smaller window can be defined to highlight region of interest.
A line plot will be displayed for both start and end of feature. If Input files are provided, ratio over Input is
computed and displayed as well."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot signals around the start and the end of genomic features</h1>
    
    <div class="hidden name"><code>plot_start_end.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot reads or peak Coverage/base/gene of samples given in the query files around start and end of custom features. The upstream and downstream windows
can be given separately, within the window, a smaller window can be defined to highlight region of interest.
A line plot will be displayed for both start and end of feature. If Input files are provided, ratio over Input is
computed and displayed as well.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_start_end</span><span class="op">(</span></span>
<span>  <span class="va">queryFiles</span>,</span>
<span>  inputFiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">centerFiles</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  importParams <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  binSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  insert <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">500</span>, <span class="fl">100</span>, <span class="op">-</span><span class="fl">100</span>, <span class="fl">500</span><span class="op">)</span>,</span>
<span>  hl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span>, <span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>  stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rmOutlier <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  shade <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span>
<span>  hw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>  nc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>queryFiles</dt>
<dd><p>a vector of sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>inputFiles</dt>
<dd><p>a vector of input sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>centerFiles</dt>
<dd><p>bed files that define the custom features, or features in c("utr3", "utr5", "cds", "intron", "exon", "transcript", "gene"), multiple features are allowed.</p></dd>


<dt>txdb</dt>
<dd><p>a TxDb object defined in the GenomicFeatures package. Default NULL, needed only when genomic features are used in the place of centerFiles.</p></dd>


<dt>importParams</dt>
<dd><p>a list of parameters for <code>handle_input</code></p></dd>


<dt>binSize</dt>
<dd><p>an integer defines bin size for intensity calculation</p></dd>


<dt>insert</dt>
<dd><p>an integer specifies the length of the center regions to be included, in addition to the start and end of the feature</p></dd>


<dt>verbose</dt>
<dd><p>logical, whether to output additional information (including data used for plotting or statistical test results)</p></dd>


<dt>ext</dt>
<dd><p>a vector of four integers defining upstream and downstream boundaries of the plot window, flanking the start and end of features</p></dd>


<dt>hl</dt>
<dd><p>a vector of four integers defining upstream and downstream boundaries of the highlight window, flanking the start and end of features</p></dd>


<dt>stranded</dt>
<dd><p>logical, indicating whether the strand of the feature should be considered</p></dd>


<dt>scale</dt>
<dd><p>logical, indicating whether the score matrix should be scaled to the range 0:1, so that samples with different baseline can be compared</p></dd>


<dt>smooth</dt>
<dd><p>logical, indicating whether the line should smoothed with a spline smoothing algorithm</p></dd>


<dt>rmOutlier</dt>
<dd><p>a numeric value serving as a multiplier of the MAD in Hampel filter for outliers identification, 0 indicating not removing outliers. For Gaussian distribution, use 3, adjust based on data distribution</p></dd>


<dt>outPrefix</dt>
<dd><p>a string specifying output file prefix for plots (outPrefix.pdf)</p></dd>


<dt>transform</dt>
<dd><p>a string in c("log", "log2", "log10"), default = NA, indicating no transformation of data matrix</p></dd>


<dt>shade</dt>
<dd><p>logical indicating whether to place a shaded rectangle around the point of interest</p></dd>


<dt>Ylab</dt>
<dd><p>a string for y-axis label</p></dd>


<dt>hw</dt>
<dd><p>a vector of two elements specifying the height and width of the output figures</p></dd>


<dt>nc</dt>
<dd><p>integer, number of cores for parallel processing</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list of two objects, the first is a GRanges object, the second is a GRangesList object</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Shuye Pu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"query"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"input"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.narrowPeak"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"narrowPeak"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">FALSE</span>, </span></span>
<span class="r-in"><span>useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">TRUE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">plot_start_end</span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, inputFiles<span class="op">=</span><span class="va">inputfiles</span>, </span></span>
<span class="r-in"><span>centerFiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="va">centerfiles</span><span class="op">)</span>, txdb<span class="op">=</span><span class="va">txdb</span>, importParams<span class="op">=</span><span class="va">importParams</span>, binSize<span class="op">=</span><span class="fl">10</span>,</span></span>
<span class="r-in"><span>insert<span class="op">=</span><span class="fl">100</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, ext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">500</span>, <span class="fl">100</span>, <span class="op">-</span><span class="fl">100</span>, <span class="fl">500</span><span class="op">)</span>, hl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span>, <span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span><span class="op">)</span>, stranded<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>scale<span class="op">=</span><span class="cn">FALSE</span>, smooth<span class="op">=</span><span class="cn">TRUE</span>, rmOutlier<span class="op">=</span><span class="fl">0</span>, outPrefix<span class="op">=</span><span class="cn">NULL</span>, transform<span class="op">=</span><span class="cn">NA</span>, shade<span class="op">=</span><span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>Ylab<span class="op">=</span><span class="st">"Coverage/base/gene"</span>, nc<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/treat_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 144852 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/treat_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/input_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 240639 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/input_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bed file: C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/test_chip_peak_chr19.narrowPeak"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The input file C:/Users/greenblatt/AppData/Local/Temp/RtmpuKLw25/temp_libpath57068a86f55/GenomicPlot/extdata/test_chip_peak_chr19.narrowPeak have more than 6 columns, only the first 6 columns will be used!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Bin numbers 60 10 60"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of features:  gene 1483"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Start 60 1483"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         77964        0.8761969</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0   67 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.05  0.05  0.05  0.05 67.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         79304        0.8912565</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0  290 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 290.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Center 10 1483"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         12597        0.8494268</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0   29 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.05  0.05  0.05  0.05 29.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         12495        0.8425489</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0   0.0   0.0   0.0 137.1 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 137.10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "End 60 1483"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         65063        0.7312093</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    1  151 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   1.00 151.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         70537        0.7927287</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0   0.0   0.0   0.0 367.4 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 367.40 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of features:  narrowPeak 5825"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Start 60 5825"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        277776        0.7947811</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0  226 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 226.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        285065        0.8156366</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0  489 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 489.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Center 10 5825"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         45705        0.7846352</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0   53 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.05  0.05  0.05  0.05 53.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         44385        0.7619742</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0  275 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 275.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "End 60 5825"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        280985        0.8039628</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0%  25%  50%  75% 100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0    0    0    0  151 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 151.00 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        285873        0.8179485</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0%   25%   50%   75%  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0   0.0   0.0   0.0 450.9 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-plt img"><img src="plot_start_end-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0%    25%    50%    75%   100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.05   0.05   0.05   0.05 450.90 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.05"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-plt img"><img src="plot_start_end-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_start_end-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_start_end-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "67 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01643489  1.00000000  1.00000000  1.00000000 16.19047619 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01643489  1.00000000  1.00000000  1.00000000 16.19047619 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.00821744627054362"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "29 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         0%        25%        50%        75%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.0362069  1.0000000  1.0000000  1.0000000 13.3333333 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         0%        25%        50%        75%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.0362069  1.0000000  1.0000000  1.0000000 13.3333333 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0181034482758621"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "151 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01779661  1.00000000  1.00000000  1.00000000 44.76190476 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01779661  1.00000000  1.00000000  1.00000000 44.76190476 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.00889830508474576"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "226 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01761745  1.00000000  1.00000000  1.00000000 51.94444444 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01761745  1.00000000  1.00000000  1.00000000 51.94444444 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.00880872483221476"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "53 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.04251012  1.00000000  1.00000000  1.00000000 38.09523810 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.04251012  1.00000000  1.00000000  1.00000000 38.09523810 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0212550607287449"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "151 0.05 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01557864  1.00000000  1.00000000  1.00000000 48.73684211 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-plt img"><img src="plot_start_end-5.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.01557864  1.00000000  1.00000000  1.00000000 48.73684211 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0077893175074184"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-plt img"><img src="plot_start_end-6.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_start_end-7.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_start_end-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span> </span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

