<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot signal inside as well as around custom genomic regions — plot_region • GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot signal inside as well as around custom genomic regions — plot_region"><meta property="og:description" content="Plot reads or peak Coverage/base/gene of samples given in the query files inside regions defined in the centerFiles. The upstream and downstream flanking windows can be given separately. If Input files are provided, ratio over Input is computed and displayed as well."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot signal inside as well as around custom genomic regions</h1>
    
    <div class="hidden name"><code>plot_region.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot reads or peak Coverage/base/gene of samples given in the query files inside regions defined in the centerFiles. The upstream and downstream flanking windows can be given separately. If Input files are provided, ratio over Input is computed and displayed as well.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_region</span><span class="op">(</span></span>
<span>  <span class="va">queryFiles</span>,</span>
<span>  <span class="va">centerFiles</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  regionName <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  inputFiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nbins <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  importParams <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fiveP <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span>,</span>
<span>  threeP <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rmOutlier <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span>
<span>  statsMethod <span class="op">=</span> <span class="st">"wilcox.test"</span>,</span>
<span>  hw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>  nc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>queryFiles</dt>
<dd><p>a named vector of sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>centerFiles</dt>
<dd><p>a named vector of reference file names or genomic features in  c("utr3", "utr5", "cds", "intron", "exon", "transcript", "gene"). The file should be in .bed format only</p></dd>


<dt>txdb</dt>
<dd><p>a TxDb object defined in the GenomicFeatures package. Default NULL, needed only when genomic features are used in the place of centerFiles.</p></dd>


<dt>regionName</dt>
<dd><p>a string specifying the name of the center region in the plots</p></dd>


<dt>inputFiles</dt>
<dd><p>a named vector of input sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>nbins</dt>
<dd><p>an integer defines the total number of bins</p></dd>


<dt>importParams</dt>
<dd><p>a list of parameters for <code>handle_input</code></p></dd>


<dt>verbose</dt>
<dd><p>logical, indicating whether to output additional information (data used for plotting or statistical test results)</p></dd>


<dt>scale</dt>
<dd><p>logical, indicating whether the score matrix should be scaled to the range 0:1, so that samples with different baseline can be compared</p></dd>


<dt>heatmap</dt>
<dd><p>logical, indicating whether a heatmap of the score matrix should be generated</p></dd>


<dt>fiveP</dt>
<dd><p>an integer, indicating extension out or inside of the 5' boundary of gene by negative or positive number</p></dd>


<dt>threeP</dt>
<dd><p>an integer, indicating extension out or inside of the 5' boundary of gene by positive or negative number</p></dd>


<dt>smooth</dt>
<dd><p>logical, indicating whether the line should smoothed with a spline smoothing algorithm</p></dd>


<dt>stranded</dt>
<dd><p>logical, indicating whether the strand of the feature should be considered</p></dd>


<dt>transform</dt>
<dd><p>a string in c("log", "log2", "log10"), default = NA indicating no transformation of data matrix</p></dd>


<dt>outPrefix</dt>
<dd><p>a string specifying output file prefix for plots (outPrefix.pdf)</p></dd>


<dt>rmOutlier</dt>
<dd><p>a numeric value serving as a multiplier of the MAD in Hampel filter for outliers identification, 0 indicating not removing outliers. For Gaussian distribution, use 3, adjust based on data distribution</p></dd>


<dt>heatRange</dt>
<dd><p>a numerical vector of two elements, defining range for heatmap color ramp generation</p></dd>


<dt>Ylab</dt>
<dd><p>a string for y-axis label</p></dd>


<dt>statsMethod</dt>
<dd><p>a string in c("wilcox.test", "t.test"), for pair-wise group comparisons</p></dd>


<dt>hw</dt>
<dd><p>a vector of two elements specifying the height and width of the output figures</p></dd>


<dt>nc</dt>
<dd><p>integer, number of cores for parallel processing</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a dataframe containing the data used for plotting</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Shuye Pu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"query"</span></span></span>
<span class="r-in"><span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"chip_input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"input"</span></span></span>
<span class="r-in"><span><span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chip_peak_chr19.narrowPeak"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"narrowPeak"</span></span></span>
<span class="r-in"><span><span class="va">op</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">TRUE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="fu">plot_region</span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, centerFiles<span class="op">=</span><span class="va">centerfiles</span>, txdb<span class="op">=</span><span class="cn">NULL</span>, regionName<span class="op">=</span><span class="st">"region"</span>, </span></span>
<span class="r-in"><span>inputFiles<span class="op">=</span><span class="va">inputfiles</span>, nbins<span class="op">=</span><span class="fl">100</span>, importParams<span class="op">=</span><span class="va">importParams</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>scale<span class="op">=</span><span class="cn">FALSE</span>, heatmap<span class="op">=</span><span class="cn">TRUE</span>, fiveP<span class="op">=</span><span class="op">-</span><span class="fl">1000</span>, threeP<span class="op">=</span><span class="fl">1000</span>, smooth<span class="op">=</span><span class="cn">TRUE</span>, stranded<span class="op">=</span><span class="cn">TRUE</span>, transform<span class="op">=</span><span class="st">"log2"</span>,</span></span>
<span class="r-in"><span>outPrefix<span class="op">=</span><span class="cn">NULL</span>, rmOutlier<span class="op">=</span><span class="fl">0</span>, heatRange<span class="op">=</span><span class="cn">NULL</span>, Ylab<span class="op">=</span><span class="st">"Coverage/base/gene"</span>, </span></span>
<span class="r-in"><span>statsMethod<span class="op">=</span><span class="st">"wilcox.test"</span>, nc<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/chip_treat_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Paired end bam file detected!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/chip_treat_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/chip_input_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Paired end bam file detected!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/chip_input_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bed file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/test_chip_peak_chr19.narrowPeak"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The input file C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/test_chip_peak_chr19.narrowPeak have more than 6 columns, only the first 6 columns will be used!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "computing coverage for Sample"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "queryLabel query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centerLabel narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName -1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName region"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName 1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "queryLabel input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centerLabel narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName -1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName region"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName 1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plotting coverage profiles"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "queryLabel query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centerLabel narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        559199        0.7038376</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.000000    0.000000    0.000000    7.028187 2030.694099 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2.772748e-02 2.772748e-02 2.772748e-02 7.028187e+00 2.030694e+03 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0277274801288241"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName -1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName region"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName 1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         90%       90.5%         91%       91.5%         92%       92.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    9.919783   11.445904   12.972024   14.400892   15.774803   17.326886 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         93%       93.5%         94%       94.5%         95%       95.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   18.909169   19.076506   19.076506   19.076506   19.076506   19.076506 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         96%       96.5%         97%       97.5%         98%       98.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   22.011353   25.245136   28.614759   28.614759   34.729024   39.620436 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         99%       99.5%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   48.922008   71.894583 2030.694099 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "queryLabel input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centerLabel narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        617165        0.7767967</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       0%      25%      50%      75%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.000    0.000    0.000    0.000 2381.269 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.0210222    0.0210222    0.0210222    0.0210222 2381.2689153 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0210222020521201"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName -1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName region"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "featureName 1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         90%       90.5%         91%       91.5%         92%       92.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    7.820259    7.820259    7.820259    7.820259    7.820259    7.820259 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         93%       93.5%         94%       94.5%         95%       95.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    7.820259    7.820259    7.820259    7.820259    7.820259    7.820259 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         96%       96.5%         97%       97.5%         98%       98.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    7.820259    8.780642   11.028571   13.450846   15.640518   15.640518 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         99%       99.5%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   15.640518   18.076007 2381.268915 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-plt img"><img src="plot_region-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "beds query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centers narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "beds input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centers narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "beds query" "beds input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centers narrowPeak"</span>
<span class="r-plt img"><img src="plot_region-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_region-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "2030.69409868277 0.0210222020521201 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.02562829   1.00656722   1.00656722   1.31911487 458.12161981 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.02562829   1.00656722   1.00656722   1.31911487 458.12161981 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.012814143024458"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plotting coverage profiles"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ratiolabel query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "centerlabel narrowPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "-1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "region"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      90%    90.5%      91%    91.5%      92%    92.5%      93%    93.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3.367550 3.367550 3.367550 3.367550 3.367550 3.367550 3.367550 3.367550 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      94%    94.5%      95%    95.5%      96%    96.5%      97%    97.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3.607585 3.825773 4.022369 4.203010 4.297422 4.297422 4.297422 4.323546 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      98%    98.5%      99%    99.5%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4.636980 4.858230 5.136948 5.575577 8.839587 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "narrowPeak"</span>
<span class="r-plt img"><img src="plot_region-4.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_region-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

