<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot promoter, 5'UTR, CDS, 3'UTR and TTS — plot_5parts_metagene • GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot promoter, 5'UTR, CDS, 3'UTR and TTS — plot_5parts_metagene"><meta property="og:description" content="Plot reads or peak Coverage/base/gene of samples given in the query files around genes. The upstream and downstream windows flanking genes can be given separately, metagene plots are generated with 5'UTR, CDS and 3'UTR segments. The length of each segments are prorated according to the median length of each segments. If Input files are provided, ratio over Input is computed and displayed as well."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot promoter, 5'UTR, CDS, 3'UTR and TTS</h1>
    
    <div class="hidden name"><code>plot_5parts_metagene.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot reads or peak Coverage/base/gene of samples given in the query files around genes. The upstream and downstream windows flanking genes can be given separately, metagene plots are generated with 5'UTR, CDS and 3'UTR segments. The length of each segments are prorated according to the median length of each segments. If Input files are provided, ratio over Input is computed and displayed as well.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_5parts_metagene</span><span class="op">(</span></span>
<span>  <span class="va">queryFiles</span>,</span>
<span>  <span class="va">gFeatures_list</span>,</span>
<span>  inputFiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  importParams <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rmOutlier <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span>
<span>  hw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>  nc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>queryFiles</dt>
<dd><p>a vector of sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>gFeatures_list</dt>
<dd><p>a list of genomic features as output of the function 'prepare_5parts_genomic_features'</p></dd>


<dt>inputFiles</dt>
<dd><p>a vector of input sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>importParams</dt>
<dd><p>a list of parameters for <code>handle_input</code></p></dd>


<dt>verbose</dt>
<dd><p>logical, indicating whether to output additional information (data used for plotting or statistical test results)</p></dd>


<dt>transform</dt>
<dd><p>logical, whether to log2 transform the matrix</p></dd>


<dt>smooth</dt>
<dd><p>logical, indicating whether the line should smoothed with a spline smoothing algorithm</p></dd>


<dt>scale</dt>
<dd><p>logical, indicating whether the score matrix should be scaled to the range 0:1, so that samples with different baseline can be compared</p></dd>


<dt>stranded</dt>
<dd><p>logical, indicating whether the strand of the feature should be considered</p></dd>


<dt>outPrefix</dt>
<dd><p>a string specifying output file prefix for plots (outPrefix.pdf)</p></dd>


<dt>heatmap</dt>
<dd><p>logical, indicating whether a heatmap of the score matrix should be generated</p></dd>


<dt>heatRange</dt>
<dd><p>a numerical vector of two elements, defining range for heatmap color ramp generation</p></dd>


<dt>rmOutlier</dt>
<dd><p>a numeric value serving as a multiplier of the MAD in Hampel filter for outliers identification, 0 indicating not removing outliers. For Gaussian distribution, use 3, adjust based on data distribution</p></dd>


<dt>Ylab</dt>
<dd><p>a string for y-axis label</p></dd>


<dt>hw</dt>
<dd><p>a vector of two elements specifying the height and width of the output figures</p></dd>


<dt>nc</dt>
<dd><p>integer, number of cores for parallel processing</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a dataframe containing the data used for plotting</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Shuye Pu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"query"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"input"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gfeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="prepare_5parts_genomic_features.html">prepare_5parts_genomic_features</a></span><span class="op">(</span><span class="va">txdb</span>, meta<span class="op">=</span><span class="cn">TRUE</span>, nbins<span class="op">=</span><span class="fl">100</span>, fiveP<span class="op">=</span><span class="op">-</span><span class="fl">1000</span>, </span></span>
<span class="r-in"><span>threeP<span class="op">=</span><span class="fl">1000</span>, longest<span class="op">=</span><span class="cn">TRUE</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="fl">0</span>, fix_width<span class="op">=</span><span class="fl">150</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span></span>
<span class="r-in"><span>   </span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">plot_5parts_metagene</span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, gFeatures<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"metagene"</span><span class="op">=</span><span class="va">gfeatures</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>inputFiles<span class="op">=</span><span class="va">inputfiles</span>, scale<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, Ylab<span class="op">=</span><span class="st">"Coverage/base/gene"</span>, </span></span>
<span class="r-in"><span>importParams<span class="op">=</span><span class="va">importParams</span>, smooth<span class="op">=</span><span class="cn">TRUE</span>,  stranded<span class="op">=</span><span class="cn">TRUE</span>, outPrefix<span class="op">=</span><span class="cn">NULL</span>, heatmap<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>rmOutlier<span class="op">=</span><span class="fl">0</span>, heatRange<span class="op">=</span><span class="cn">NULL</span>, transform<span class="op">=</span><span class="st">"log2"</span>, nc<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/treat_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is used</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/input_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is used</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "computing coverage for query files in metagene"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of features:"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -1K 5'UTR   CDS 3'UTR    1K </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1328  1328  1328  1328  1328 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of scaled bins"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -1K 5'UTR   CDS 3'UTR    1K </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    25     5    30    15    25 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "-1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "5'UTR"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CDS"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "3'UTR"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "-1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "5'UTR"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CDS"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "3'UTR"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1K"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Preparing data for individual plotting"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         92852        0.6991867</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.000000    0.000000    0.000000    5.782597 1546.405987 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2.054642e-02 2.054642e-02 2.054642e-02 5.782597e+00 1.546406e+03 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0205464230814243"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        90%      90.5%        91%      91.5%        92%      92.5%        93% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   22.32739   24.34529   26.33655   27.61439   27.61439   28.66513   31.41137 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      93.5%        94%      94.5%        95%      95.5%        96%      96.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   34.39015   34.51799   36.95455   40.92905   41.42348   46.47800   49.01555 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        97%      97.5%        98%      98.5%        99%      99.5%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   55.22879   61.30403   68.11561   77.89153   96.65037  129.61566 1546.40599 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         94719        0.7132455</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.000000    0.000000    0.000000    2.585708 2011.731307 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    0.0101853    0.0101853    0.0101853    2.5857081 2011.7313075 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0101852999395645"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        90%      90.5%        91%      91.5%        92%      92.5%        93% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   13.42579   14.81029   16.23305   16.62241   17.08414   18.89030   20.77801 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      93.5%        94%      94.5%        95%      95.5%        96%      96.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   20.77801   22.77907   24.93361   26.20412   29.08922   31.14510   33.24482 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        97%      97.5%        98%      98.5%        99%      99.5%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   37.40042   41.55602   47.89401   55.32474   70.42364  107.94229 2011.73131 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Preparing data for ratio plotting"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1546.40598680032 0.0101852999395645 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4.010136e-03 1.010257e+00 1.010257e+00 1.010257e+00 3.048130e+02 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4.010136e-03 1.010257e+00 1.010257e+00 1.010257e+00 3.048130e+02 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.00200506783333769"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      90%    90.5%      91%    91.5%      92%    92.5%      93%    93.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2.967890 2.967890 2.967890 2.967890 2.967890 3.088703 3.277441 3.490686 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      94%    94.5%      95%    95.5%      96%    96.5%      97%    97.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3.684233 3.873607 3.873607 3.873607 4.015833 4.285837 4.425721 4.475830 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      98%    98.5%      99%    99.5%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4.824049 4.924864 5.135858 5.609633 8.251781 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Start plotting"</span>
<span class="r-plt img"><img src="plot_5parts_metagene-1.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_5parts_metagene-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_5parts_metagene-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_5parts_metagene-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plot_5parts_metagene runs successfully!"</span>
<span class="r-in"><span>               </span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

