<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot promoter, gene body and TTS — plot_3parts_metagene • GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot promoter, gene body and TTS — plot_3parts_metagene"><meta property="og:description" content="Plot reads or peak Coverage/base/gene of samples given in the query files around genes. The upstream and downstream windows flanking genes
can be given separately, the parameter 'meta' controls if gene or metagene plots are generated. If Input files are provided, ratio over Input
is computed and displayed as well."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/shuye2009/GenomicPlot/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot promoter, gene body and TTS</h1>
    <small class="dont-index">Source: <a href="https://github.com/shuye2009/GenomicPlot/blob/HEAD/R/GenomicPlot.R" class="external-link"><code>R/GenomicPlot.R</code></a></small>
    <div class="hidden name"><code>plot_3parts_metagene.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot reads or peak Coverage/base/gene of samples given in the query files around genes. The upstream and downstream windows flanking genes
can be given separately, the parameter 'meta' controls if gene or metagene plots are generated. If Input files are provided, ratio over Input
is computed and displayed as well.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_3parts_metagene</span><span class="op">(</span></span>
<span>  <span class="va">queryFiles</span>,</span>
<span>  <span class="va">gFeatures</span>,</span>
<span>  inputFiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span>
<span>  importParams <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rmOutlier <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  hw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>  nc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>queryFiles</dt>
<dd><p>a vector of sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>gFeatures</dt>
<dd><p>genomic features as output of the function 'prepare_3parts_genomic_features'</p></dd>


<dt>inputFiles</dt>
<dd><p>a vector of input sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>scale</dt>
<dd><p>logical, indicating whether the score matrix should be scaled to the range 0:1, so that samples with different baseline can be compared</p></dd>


<dt>verbose</dt>
<dd><p>logical, whether to output additional information (data used for plotting or statistical test results)</p></dd>


<dt>Ylab</dt>
<dd><p>a string for y-axis label</p></dd>


<dt>importParams</dt>
<dd><p>a list of parameters for <code>handle_input</code></p></dd>


<dt>smooth</dt>
<dd><p>logical, indicating whether the line should smoothed with a spline smoothing algorithm</p></dd>


<dt>stranded</dt>
<dd><p>logical, indicating whether the strand of the feature should be considered</p></dd>


<dt>outPrefix</dt>
<dd><p>a string specifying output file prefix for plots (outPrefix.pdf)</p></dd>


<dt>heatmap</dt>
<dd><p>logical, indicating whether a heatmap of the score matrix should be generated</p></dd>


<dt>rmOutlier</dt>
<dd><p>a numeric value serving as a multiplier of the MAD in Hampel filter for outliers identification, 0 indicating not removing outliers. For Gaussian distribution, use 3, adjust based on data distribution</p></dd>


<dt>heatRange</dt>
<dd><p>a numerical vector of two elements, defining range for heatmap color ramp generation</p></dd>


<dt>transform</dt>
<dd><p>a string in c("log", "log2", "log10"), default = NA indicating no transformation of data matrix</p></dd>


<dt>hw</dt>
<dd><p>a vector of two elements specifying the height and width of the output figures</p></dd>


<dt>nc</dt>
<dd><p>integer, number of cores for parallel processing</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a dataframe containing the data used for plotting</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Shuye Pu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"GenomicPlot"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"query"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package <span class="op">=</span> <span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"input"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gfeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="prepare_3parts_genomic_features.html">prepare_3parts_genomic_features</a></span><span class="op">(</span><span class="va">txdb</span>,</span></span>
<span class="r-in"><span>  featureName <span class="op">=</span> <span class="st">"transcript"</span>, meta <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  nbins <span class="op">=</span> <span class="fl">100</span>, fiveP <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span>, threeP <span class="op">=</span> <span class="fl">1000</span>, longest <span class="op">=</span> <span class="cn">TRUE</span>, protein_coding <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  offset <span class="op">=</span> <span class="fl">0</span>, fix_width <span class="op">=</span> <span class="fl">150</span>, fix_point <span class="op">=</span> <span class="st">"start"</span>, norm <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  useScore <span class="op">=</span> <span class="cn">FALSE</span>, outRle <span class="op">=</span> <span class="cn">TRUE</span>, useSizeFactor <span class="op">=</span> <span class="cn">FALSE</span>, genome <span class="op">=</span> <span class="st">"hg19"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">plot_3parts_metagene</span><span class="op">(</span></span></span>
<span class="r-in"><span>  queryFiles <span class="op">=</span> <span class="va">queryfiles</span>, gFeatures <span class="op">=</span> <span class="va">gfeatures</span>,</span></span>
<span class="r-in"><span>  inputFiles <span class="op">=</span> <span class="va">inputfiles</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span></span>
<span class="r-in"><span>  importParams <span class="op">=</span> <span class="va">importParams</span>, smooth <span class="op">=</span> <span class="cn">TRUE</span>, stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>, heatmap <span class="op">=</span> <span class="cn">TRUE</span>, rmOutlier <span class="op">=</span> <span class="fl">0</span>, heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  transform <span class="op">=</span> <span class="st">"log2"</span>, nc <span class="op">=</span> <span class="fl">2</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading bamfile: C:/Users/greenblatt/AppData/Local/Temp/RtmpiQE4hr/temp_libpath167c637a5c73/GenomicPlot/extdata/treat_chr19.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 144852 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpiQE4hr/temp_libpath167c637a5c73/GenomicPlot/extdata/treat_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cached .rds file is modified using new input parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading bamfile: C:/Users/greenblatt/AppData/Local/Temp/RtmpiQE4hr/temp_libpath167c637a5c73/GenomicPlot/extdata/input_chr19.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 240639 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpiQE4hr/temp_libpath167c637a5c73/GenomicPlot/extdata/input_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cached .rds file is modified using new input parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K transcript 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Window sizes: 1485 1485 1485</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing coverage for queryFiles...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: transcript</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: transcript</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Plotting coverage for queryFiles...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inspecting matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero         27652        0.8096033</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quartiles:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0 0 0 0 2314.67052306567</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.0472227542652536 0.0472227542652536 0.0472227542652536 0.0472227542652536 2314.67052306567</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The imputed value is: 0.0472227542652536</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: transcript</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Drawing heatmap</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 23.5349937114497 25.3712147147873 27.293295032813 27.6143926214343 27.7650165811877 30.2816919087319 32.7648863808158 34.5179907767929 35.2446853194622 38.5973896867775 41.4215889321514 43.5390611484143 48.32518708751 50.8355864167313 55.3482248157288 62.1323833982272 69.0359815535857 79.7927507724003 96.65037417502 130.088350125008 1546.40598680032</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inspecting matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quartiles:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.0472227542652536 0.0472227542652536 0.0472227542652536 0.0472227542652536 2314.67052306567</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.0472227542652536 0.0472227542652536 0.0472227542652536 0.0472227542652536 2314.67052306567</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The imputed value is: 0.0236113771326268</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: transcript</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Drawing heatmap</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 13.6697446557313 14.8828550186678 16.1464517753645 16.6224095013693 17.1376306454273 18.8079931502921 20.626042142839 20.7780118767116 22.5365263653335 24.9336142520539 25.9966753248159 28.7540874035783 30.6003447638843 33.2448190027385 37.4004213780809 41.5560237534232 46.5640865780972 54.5563610133629 67.4344264278714 104.14547862062 2314.67052306567</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-plt img"><img src="plot_3parts_metagene-1.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_3parts_metagene-2.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing coverage for ratio over input...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Max(IP) min(Input)     pseudo </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1546.406      0.000      1.000 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inspecting matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quartiles:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.00392940054070958 1 1 1 312.352276806672</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.00392940054070958 1 1 1 312.352276806672</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The imputed value is: 0.00196470027035479</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Plotting coverage for ratio over input...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: -1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: transcript</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Feature name: 1K</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Drawing heatmap</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2.98250959747812 2.98250959747812 2.98250959747812 2.98250959747812 3.00737712879301 3.1779080925056 3.35855064098628 3.54465305388625 3.726634710542 3.88822659226355 3.88822659226355 3.88822659226355 4.0575196555014 4.31470593118786 4.44034061454361 4.52325988811011 4.83866907897695 4.97530529142935 5.2507970209593 5.62425261790332 8.28703023584823</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>implicit list embedding of S4 objects is deprecated</span>
<span class="r-plt img"><img src="plot_3parts_metagene-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_3parts_metagene-4.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_3parts_metagene-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

