<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot signal around custom genomic loci — plot_locus • GenomicPlot</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot signal around custom genomic loci — plot_locus"><meta property="og:description" content="Plot reads or peak Coverage/base/gene of samples given in the query files around reference locus (start, end or center of a genomic region)
defined in the centerFiles. The upstream and downstream windows flanking loci can be given separately, a smaller window can be defined to allow
statistical comparisons between samples for the same reference, or between references for a given sample. If Input files are provided, ratio over
Input is computed and displayed as well."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicPlot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicPlot_vignettes.html">GenomicPlot: an R package for efficient and flexible visualization of NGS coverage profiles</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot signal around custom genomic loci</h1>
    
    <div class="hidden name"><code>plot_locus.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot reads or peak Coverage/base/gene of samples given in the query files around reference locus (start, end or center of a genomic region)
defined in the centerFiles. The upstream and downstream windows flanking loci can be given separately, a smaller window can be defined to allow
statistical comparisons between samples for the same reference, or between references for a given sample. If Input files are provided, ratio over
Input is computed and displayed as well.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_locus</span><span class="op">(</span></span>
<span>  <span class="va">queryFiles</span>,</span>
<span>  <span class="va">centerFiles</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  hl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  shade <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  importParams <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  binSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  refPoint <span class="op">=</span> <span class="st">"center"</span>,</span>
<span>  Xlab <span class="op">=</span> <span class="st">"Center"</span>,</span>
<span>  Ylab <span class="op">=</span> <span class="st">"Coverage/base/gene"</span>,</span>
<span>  inputFiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stranded <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rmOutlier <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  statsMethod <span class="op">=</span> <span class="st">"wilcox.test"</span>,</span>
<span>  heatRange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  hw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>  nc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>queryFiles</dt>
<dd><p>a vector of sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>centerFiles</dt>
<dd><p>a named vector of reference file names or genomic features in  c("utr3", "utr5", "cds", "intron", "exon", "transcript", "gene"). The file should be in .bed format only</p></dd>


<dt>txdb</dt>
<dd><p>a TxDb object defined in the GenomicFeatures package. Default NULL, needed only when genomic features are used in the place of centerFiles.</p></dd>


<dt>ext</dt>
<dd><p>a vector of two integers defining upstream and downstream boundaries of the plot window, flanking the reference locus</p></dd>


<dt>hl</dt>
<dd><p>a vector of two integers defining upstream and downstream boundaries of the highlight window, flanking the reference locus</p></dd>


<dt>shade</dt>
<dd><p>logical indicating whether to place a shaded rectangle around the point of interest</p></dd>


<dt>smooth</dt>
<dd><p>logical, indicating whether the line should smoothed with a spline smoothing algorithm</p></dd>


<dt>importParams</dt>
<dd><p>a list of parameters for <code>handle_input</code></p></dd>


<dt>verbose</dt>
<dd><p>logical, indicating whether to output additional information (data used for plotting or statistical test results)</p></dd>


<dt>binSize</dt>
<dd><p>an integer defines bin size for intensity calculation</p></dd>


<dt>refPoint</dt>
<dd><p>a string in c("start", "center", "end")</p></dd>


<dt>Xlab</dt>
<dd><p>a string denotes the label on x-axis</p></dd>


<dt>Ylab</dt>
<dd><p>a string for y-axis label</p></dd>


<dt>inputFiles</dt>
<dd><p>a vector of input sample file names. The file should be in .bam, .bed, .wig or .bw format, mixture of formats is allowed</p></dd>


<dt>stranded</dt>
<dd><p>logical, indicating whether the strand of the feature should be considered</p></dd>


<dt>heatmap</dt>
<dd><p>logical, indicating whether a heatmap of the score matrix should be generated</p></dd>


<dt>scale</dt>
<dd><p>logical, indicating whether the score matrix should be scaled to the range 0:1, so that samples with different baseline can be compared</p></dd>


<dt>outPrefix</dt>
<dd><p>a string specifying output file prefix for plots (outPrefix.pdf)</p></dd>


<dt>rmOutlier</dt>
<dd><p>a numeric value serving as a multiplier of the MAD in Hampel filter for outliers identification, 0 indicating not removing outliers. For Gaussian distribution, use 3, adjust based on data distribution.</p></dd>


<dt>transform</dt>
<dd><p>a string in c("log", "log2", "log10"), default = NA indicating no transformation of data matrix</p></dd>


<dt>statsMethod</dt>
<dd><p>a string in c("wilcox.test", "t.test"), for pair-wise group comparisons</p></dd>


<dt>heatRange</dt>
<dd><p>a numerical vector of two elements, defining range for heatmap color ramp generation</p></dd>


<dt>hw</dt>
<dd><p>a vector of two elements specifying the height and width of the output figures</p></dd>


<dt>nc</dt>
<dd><p>integer, number of cores for parallel processing</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list of two dataframes containing the data used for plotting and for statistical testing</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Shuye Pu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"txdb_chr19.sql"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">queryfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"treat_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"query"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">inputfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input_chr19.bam"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inputfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"input"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">centerfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_clip_peak_chr19.bed"</span>, package<span class="op">=</span><span class="st">"GenomicPlot"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">centerfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clipPeak"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">importParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>offset<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, fix_width<span class="op">=</span><span class="fl">0</span>, fix_point<span class="op">=</span><span class="st">"start"</span>, norm<span class="op">=</span><span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>useScore<span class="op">=</span><span class="cn">FALSE</span>, outRle<span class="op">=</span><span class="cn">TRUE</span>, useSizeFactor<span class="op">=</span><span class="cn">FALSE</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">plot_locus</span><span class="op">(</span>queryFiles<span class="op">=</span><span class="va">queryfiles</span>, centerFiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">centerfiles</span>, <span class="st">"exon"</span><span class="op">)</span>, txdb<span class="op">=</span><span class="va">txdb</span>, </span></span>
<span class="r-in"><span>ext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">200</span>,<span class="fl">200</span><span class="op">)</span>, hl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">50</span>, <span class="fl">50</span><span class="op">)</span>, shade<span class="op">=</span><span class="cn">TRUE</span>, smooth<span class="op">=</span><span class="cn">FALSE</span>, importParams<span class="op">=</span><span class="va">importParams</span>, </span></span>
<span class="r-in"><span>verbose<span class="op">=</span><span class="cn">TRUE</span>, binSize<span class="op">=</span><span class="fl">10</span>, refPoint<span class="op">=</span><span class="st">"center"</span>, Xlab<span class="op">=</span><span class="st">"Center"</span>, Ylab<span class="op">=</span><span class="st">"Coverage/base/gene"</span>, </span></span>
<span class="r-in"><span>inputFiles<span class="op">=</span><span class="va">inputfiles</span>, stranded<span class="op">=</span><span class="cn">TRUE</span>, heatmap<span class="op">=</span><span class="cn">TRUE</span>, scale<span class="op">=</span><span class="cn">FALSE</span>, outPrefix<span class="op">=</span><span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>rmOutlier<span class="op">=</span><span class="fl">0</span>, transform<span class="op">=</span><span class="st">"log2"</span>, statsMethod<span class="op">=</span><span class="st">"wilcox.test"</span>, heatRange<span class="op">=</span><span class="cn">NULL</span>, nc<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/treat_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 144852 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/treat_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bam file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/input_chr19.bam"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 240639 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bam file C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/input_chr19.bam is loaded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is modified using new input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Reading bed file: C:/Users/greenblatt/AppData/Local/Temp/RtmpqQgWpD/temp_libpath58243c1a4510/GenomicPlot/extdata/test_clip_peak_chr19.bed"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cached .rds file is used</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "computing coverage for Sample"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "size of query regions 144852"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "preparing window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of window regions 15717"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        581335        0.9246914</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       0%      25%      50%      75%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0000   0.0000   0.0000   0.0000 541.2421 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.3451799   0.3451799   0.3451799   0.3451799 541.2420954 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.345179907767929"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Dimension of fullMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 15717    40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of unique window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "preparing window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of window regions 12403"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        480128        0.9677659</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       0%      25%      50%      75%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.00000  0.00000  0.00000  0.00000 30.37583 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         0%        25%        50%        75%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.3451799  0.3451799  0.3451799  0.3451799 30.3758319 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.345179907767929"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Dimension of fullMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 12403    40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of unique window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "size of query regions 240639"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "preparing window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of window regions 15717"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0        0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        587220        0.9340523</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       0%      25%      50%      75%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0000   0.0000   0.0000   0.0000 796.2134 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.2077801   0.2077801   0.2077801 796.2134151 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.207780118767116"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Dimension of fullMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 15717    40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of unique window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "preparing window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "number of window regions 12403"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero        476430         0.960312</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       0%      25%      50%      75%     100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.0000   0.0000   0.0000   0.0000 170.3797 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.2077801   0.2077801   0.2077801 170.3796974 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.207780118767116"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Dimension of fullMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 12403    40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Number of unique window regions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "collecting coverage data"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         90%       90.5%         91%       91.5%         92%       92.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.3451799   0.3451799   0.3451799   0.3451799   0.3451799   0.6903598 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         93%       93.5%         94%       94.5%         95%       95.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.6903598   0.6903598   0.6903598   0.6903598   0.6903598   0.6903598 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         96%       96.5%         97%       97.5%         98%       98.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1.3807196   1.3807196   1.3807196   2.0710794   2.0710794   2.7614393 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         99%       99.5%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2.7614393   4.1421589 541.2420954 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        90%      90.5%        91%      91.5%        92%      92.5%        93% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      93.5%        94%      94.5%        95%      95.5%        96%      96.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799  0.3451799 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        97%      97.5%        98%      98.5%        99%      99.5%       100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.6903598  0.6903598  0.6903598  1.3807196  1.3807196  2.7614393 30.3758319 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         90%       90.5%         91%       91.5%         92%       92.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         93%       93.5%         94%       94.5%         95%       95.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.4155602   0.4155602   0.4155602   0.4155602   0.4155602 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         96%       96.5%         97%       97.5%         98%       98.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.4155602   0.4155602   0.8311205   0.8311205   0.8311205   1.2466807 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         99%       99.5%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1.6622410   2.4933614 796.2134151 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         90%       90.5%         91%       91.5%         92%       92.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         93%       93.5%         94%       94.5%         95%       95.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801   0.2077801 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         96%       96.5%         97%       97.5%         98%       98.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.2077801   0.4155602   0.4155602   0.4155602   0.8311205   0.8311205 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         99%       99.5%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1.2466807   2.0778012 170.3796974 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-plt img"><img src="plot_locus-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Plotting profile and boxplot"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak" "exon"    </span>
<span class="r-plt img"><img src="plot_locus-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_locus-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_locus-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak" "exon"    </span>
<span class="r-plt img"><img src="plot_locus-5.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_locus-6.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Computing Ratio over input"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "541.242095380112 0.207780118767116 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.009091656  1.113762254  1.113762254  1.113762254 11.435695573 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           0%          25%          50%          75%         100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.009091656  1.113762254  1.113762254  1.113762254 11.435695573 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.00454582796247873"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "30.3758318835777 0.207780118767116 1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Inspecting matrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          n_invalid fraction_invalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> infinite         0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA               0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NaN              0                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> zero             0                0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing values. Matrix quartiles:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.02599262  1.11376225  1.11376225  1.11376225 11.11665648 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Matrix quantiles after imputing:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0.02599262  1.11376225  1.11376225  1.11376225 11.11665648 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The imputed value is: 0.0129963108442878"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "collecting ratio data"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       90%     90.5%       91%     91.5%       92%     92.5%       93%     93.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.2559572 0.4849725 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       94%     94.5%       95%     95.5%       96%     96.5%       97%     97.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.4849725 0.4849725 0.4849725 0.4849725 0.7500246 0.9790399 0.9790399 0.9790399 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       98%     98.5%       99%     99.5%      100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1.3463880 1.3463880 1.6389270 1.8820307 3.5154722 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "drawing heatmap"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "quantile(fullMatrix, c(seq(0.9, 1, 0.005)), na.rm=TRUE)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       90%     90.5%       91%     91.5%       92%     92.5%       93%     93.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       94%     94.5%       95%     95.5%       96%     96.5%       97%     97.5% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.1554413 0.4849725 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       98%     98.5%       99%     99.5%      100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.4849725 0.9790399 0.9790399 1.3463880 3.4746510 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> TRUE/FALSE to it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-plt img"><img src="plot_locus-7.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Plotting ratio profile and boxplot"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "exon"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "query"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "clipPeak" "exon"    </span>
<span class="r-plt img"><img src="plot_locus-8.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_locus-9.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="plot_locus-10.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Shuye Pu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

